// react-notion-x/build/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// ../../node_modules/.pnpm/lodash.throttle@4.1.1/node_modules/lodash.throttle/index.js
var require_lodash = __commonJS({
  "../../node_modules/.pnpm/lodash.throttle@4.1.1/node_modules/lodash.throttle/index.js"(exports, module) {
    "use strict";
    var FUNC_ERROR_TEXT = "Expected a function";
    var NAN = 0 / 0;
    var symbolTag = "[object Symbol]";
    var reTrim = /^\s+|\s+$/g;
    var reIsBadHex = /^[-+]0x[0-9a-f]+$/i;
    var reIsBinary = /^0b[01]+$/i;
    var reIsOctal = /^0o[0-7]+$/i;
    var freeParseInt = parseInt;
    var freeGlobal = typeof global == "object" && global && global.Object === Object && global;
    var freeSelf = typeof self == "object" && self && self.Object === Object && self;
    var root = freeGlobal || freeSelf || Function("return this")();
    var objectProto = Object.prototype;
    var objectToString = objectProto.toString;
    var nativeMax = Math.max;
    var nativeMin = Math.min;
    var now = function() {
      return root.Date.now();
    };
    function debounce(func, wait, options) {
      var lastArgs, lastThis, maxWait, result, timerId, lastCallTime, lastInvokeTime = 0, leading = false, maxing = false, trailing = true;
      if (typeof func != "function") {
        throw new TypeError(FUNC_ERROR_TEXT);
      }
      wait = toNumber(wait) || 0;
      if (isObject(options)) {
        leading = !!options.leading;
        maxing = "maxWait" in options;
        maxWait = maxing ? nativeMax(toNumber(options.maxWait) || 0, wait) : maxWait;
        trailing = "trailing" in options ? !!options.trailing : trailing;
      }
      function invokeFunc(time) {
        var args = lastArgs, thisArg = lastThis;
        lastArgs = lastThis = void 0;
        lastInvokeTime = time;
        result = func.apply(thisArg, args);
        return result;
      }
      function leadingEdge(time) {
        lastInvokeTime = time;
        timerId = setTimeout(timerExpired, wait);
        return leading ? invokeFunc(time) : result;
      }
      function remainingWait(time) {
        var timeSinceLastCall = time - lastCallTime, timeSinceLastInvoke = time - lastInvokeTime, result2 = wait - timeSinceLastCall;
        return maxing ? nativeMin(result2, maxWait - timeSinceLastInvoke) : result2;
      }
      function shouldInvoke(time) {
        var timeSinceLastCall = time - lastCallTime, timeSinceLastInvoke = time - lastInvokeTime;
        return lastCallTime === void 0 || timeSinceLastCall >= wait || timeSinceLastCall < 0 || maxing && timeSinceLastInvoke >= maxWait;
      }
      function timerExpired() {
        var time = now();
        if (shouldInvoke(time)) {
          return trailingEdge(time);
        }
        timerId = setTimeout(timerExpired, remainingWait(time));
      }
      function trailingEdge(time) {
        timerId = void 0;
        if (trailing && lastArgs) {
          return invokeFunc(time);
        }
        lastArgs = lastThis = void 0;
        return result;
      }
      function cancel() {
        if (timerId !== void 0) {
          clearTimeout(timerId);
        }
        lastInvokeTime = 0;
        lastArgs = lastCallTime = lastThis = timerId = void 0;
      }
      function flush() {
        return timerId === void 0 ? result : trailingEdge(now());
      }
      function debounced() {
        var time = now(), isInvoking = shouldInvoke(time);
        lastArgs = arguments;
        lastThis = this;
        lastCallTime = time;
        if (isInvoking) {
          if (timerId === void 0) {
            return leadingEdge(lastCallTime);
          }
          if (maxing) {
            timerId = setTimeout(timerExpired, wait);
            return invokeFunc(lastCallTime);
          }
        }
        if (timerId === void 0) {
          timerId = setTimeout(timerExpired, wait);
        }
        return result;
      }
      debounced.cancel = cancel;
      debounced.flush = flush;
      return debounced;
    }
    function throttle3(func, wait, options) {
      var leading = true, trailing = true;
      if (typeof func != "function") {
        throw new TypeError(FUNC_ERROR_TEXT);
      }
      if (isObject(options)) {
        leading = "leading" in options ? !!options.leading : leading;
        trailing = "trailing" in options ? !!options.trailing : trailing;
      }
      return debounce(func, wait, {
        "leading": leading,
        "maxWait": wait,
        "trailing": trailing
      });
    }
    function isObject(value) {
      var type = typeof value;
      return !!value && (type == "object" || type == "function");
    }
    function isObjectLike(value) {
      return !!value && typeof value == "object";
    }
    function isSymbol(value) {
      return typeof value == "symbol" || isObjectLike(value) && objectToString.call(value) == symbolTag;
    }
    function toNumber(value) {
      if (typeof value == "number") {
        return value;
      }
      if (isSymbol(value)) {
        return NAN;
      }
      if (isObject(value)) {
        var other = typeof value.valueOf == "function" ? value.valueOf() : value;
        value = isObject(other) ? other + "" : other;
      }
      if (typeof value != "string") {
        return value === 0 ? value : +value;
      }
      value = value.replace(reTrim, "");
      var isBinary = reIsBinary.test(value);
      return isBinary || reIsOctal.test(value) ? freeParseInt(value.slice(2), isBinary ? 2 : 8) : reIsBadHex.test(value) ? NAN : +value;
    }
    module.exports = throttle3;
  }
});

// src/components/button.tsx
import React16 from "react";

// src/context.tsx
import "notion-types";
import { defaultMapImageUrl, defaultMapPageUrl } from "notion-utils";
import React15 from "react";

// src/components/asset-wrapper.tsx
import "notion-types";
import { parsePageId as parsePageId2 } from "notion-utils";

// src/utils.ts
import { formatDate, formatNotionDateTime, isUrl } from "notion-utils";
var cs = (...classes) => classes.filter((a) => !!a).join(" ");
var getHashFragmentValue = (url) => {
  return url.includes("#") ? url.replace(/^.+(#.+)$/, "$1") : "";
};
var isBrowser = !!globalThis.window;
var youtubeDomains = /* @__PURE__ */ new Set([
  "youtu.be",
  "youtube.com",
  "www.youtube.com",
  "youtube-nocookie.com",
  "www.youtube-nocookie.com"
]);
var getYoutubeId = (url) => {
  var _a;
  try {
    const { hostname } = new URL(url);
    if (!youtubeDomains.has(hostname)) {
      return null;
    }
    const regExp = /^.*(youtu\.be\/|v\/|u\/\w\/|embed\/|watch\?v=|&v=)([^#&?]*).*/i;
    const match = url.match(regExp);
    if (match && ((_a = match[2]) == null ? void 0 : _a.length) === 11) {
      return match[2];
    }
  } catch (e) {
  }
  return null;
};
var getUrlParams = (url) => {
  try {
    const { searchParams } = new URL(url);
    const result = {};
    for (const [key, value] of searchParams.entries()) {
      result[key] = value;
    }
    return result;
  } catch (e) {
  }
  return;
};

// src/components/asset.tsx
import "notion-types";
import { getTextContent } from "notion-utils";

// src/components/lazy-image.tsx
import { normalizeUrl } from "notion-utils";
import React from "react";

// src/components/lazy-image-full.tsx
import { Component } from "react";
import { InView } from "react-intersection-observer";
import { ofType, unionize } from "unionize";
import { jsx } from "react/jsx-runtime";
var LazyImageFullState = unionize({
  NotAsked: {},
  Buffering: {},
  // Could try to make it Promise<HTMLImageElement>,
  // but we don't use the element anyway, and we cache promises
  Loading: {},
  LoadSuccess: {},
  LoadError: ofType()
});
var Action = unionize({
  ViewChanged: ofType(),
  BufferingEnded: {},
  // MAYBE: Load: {},
  LoadSuccess: {},
  LoadError: ofType()
});
var getBufferingCmd = (durationMs) => (instance) => {
  const bufferingPromise = makeCancelable(delayedPromise(durationMs));
  bufferingPromise.promise.then(() => instance.update(Action.BufferingEnded())).catch(
    (_err) => {
    }
    //console.log({ isCanceled: _reason.isCanceled })
  );
  instance.promiseCache.buffering = bufferingPromise;
};
var getLoadingCmd = (imageProps, experimentalDecode) => (instance) => {
  const loadingPromise = makeCancelable(
    loadImage(imageProps, experimentalDecode)
  );
  loadingPromise.promise.then((_res) => instance.update(Action.LoadSuccess({}))).catch((err) => {
    if (!err.isCanceled) {
      instance.update(new Action.LoadError({ msg: "LoadError" }));
    }
  });
  instance.promiseCache.loading = loadingPromise;
};
var cancelBufferingCmd = (instance) => {
  var _a;
  (_a = instance.promiseCache.buffering) == null ? void 0 : _a.cancel();
};
var _LazyImageFull = class _LazyImageFull extends Component {
  constructor(props) {
    super(props);
    /** A central place to store promises.
     * A bit silly, but passing promises directly in the state
     * was giving me weird timing issues. This way we can keep
     * the promises in check, and pick them up from the respective methods.
     * FUTURE: Could pass the relevant key in Buffering and Loading, so
     * that at least we know where they are from a single source.
     */
    __publicField(this, "promiseCache", {});
    __publicField(this, "initialState", LazyImageFullState.NotAsked());
    this.state = this.initialState;
    this.update = this.update.bind(this);
  }
  /** Emit the next state based on actions.
   *  This is the core of the component!
   */
  static reducer(action, prevState, props) {
    return Action.match(action, {
      ViewChanged: ({ inView }) => {
        if (inView === true) {
          if (!props.src) {
            return { nextState: LazyImageFullState.LoadSuccess() };
          } else {
            return LazyImageFullState.match(prevState, {
              NotAsked: () => {
                if (props.debounceDurationMs) {
                  return {
                    nextState: LazyImageFullState.Buffering(),
                    cmd: getBufferingCmd(props.debounceDurationMs)
                  };
                } else {
                  return {
                    nextState: LazyImageFullState.Loading(),
                    cmd: getLoadingCmd(props, props.experimentalDecode)
                  };
                }
              },
              // Do nothing in other states
              default: () => ({ nextState: prevState })
            });
          }
        } else {
          return LazyImageFullState.match(prevState, {
            Buffering: () => ({
              nextState: LazyImageFullState.NotAsked(),
              cmd: cancelBufferingCmd
            }),
            // Do nothing in other states
            default: () => ({ nextState: prevState })
          });
        }
      },
      // Buffering has ended/succeeded, kick off request for image
      BufferingEnded: () => ({
        nextState: LazyImageFullState.Loading(),
        cmd: getLoadingCmd(props, props.experimentalDecode)
      }),
      // Loading the image succeeded, simple
      LoadSuccess: () => ({ nextState: LazyImageFullState.LoadSuccess() }),
      //@ts-expect-error No need for changing structure
      LoadError: (e) => ({ nextState: new LazyImageFullState.LoadError(e) })
    });
  }
  update(action) {
    const { nextState, cmd } = _LazyImageFull.reducer(
      action,
      this.state,
      this.props
    );
    if (this.props.debugActions) {
      if (false) {
        console.warn(
          'You are running LazyImage with debugActions="true" in production. This might have performance implications.'
        );
      }
      console.log({ action, prevState: this.state, nextState });
    }
    this.setState(nextState, () => cmd && cmd(this));
  }
  componentWillUnmount() {
    if (this.promiseCache.loading) {
      this.promiseCache.loading.cancel();
    }
    if (this.promiseCache.buffering) {
      this.promiseCache.buffering.cancel();
    }
    this.promiseCache = {};
  }
  // Render function
  render() {
    const { children, loadEagerly, observerProps, ...imageProps } = this.props;
    if (loadEagerly) {
      return children({
        // We know that the state tags and the enum match up
        imageState: LazyImageFullState.LoadSuccess().tag,
        imageProps
      });
    } else {
      return /* @__PURE__ */ jsx(
        InView,
        {
          rootMargin: "50px 0px",
          threshold: 0.01,
          ...observerProps,
          onChange: (inView) => this.update(Action.ViewChanged({ inView })),
          children: ({ ref }) => children({
            // We know that the state tags and the enum match up, apart
            // from Buffering not being exposed
            imageState: this.state.tag === "Buffering" ? "Loading" /* Loading */ : this.state.tag,
            imageProps,
            ref
          })
        }
      );
    }
  }
};
__publicField(_LazyImageFull, "displayName", "LazyImageFull");
var LazyImageFull = _LazyImageFull;
var loadImage = ({ src, srcSet, alt, sizes }, experimentalDecode = false) => new Promise((resolve, reject) => {
  const image = new Image();
  if (srcSet) {
    image.srcset = srcSet;
  }
  if (alt) {
    image.alt = alt;
  }
  if (sizes) {
    image.sizes = sizes;
  }
  image.src = src;
  if (experimentalDecode && "decode" in image) {
    return image.decode().then(() => resolve(image)).catch((err) => reject(err));
  }
  image.addEventListener("load", resolve);
  image.addEventListener("error", reject);
});
var delayedPromise = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
var makeCancelable = (promise) => {
  let hasCanceled_ = false;
  const wrappedPromise = new Promise((resolve, reject) => {
    void promise.then(
      (val) => hasCanceled_ ? reject({ isCanceled: true }) : resolve(val)
    );
    promise.catch(
      (err) => hasCanceled_ ? reject({ isCanceled: true }) : reject(err)
    );
  });
  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled_ = true;
    }
  };
};

// src/components/lazy-image.tsx
import { jsx as jsx2, jsxs } from "react/jsx-runtime";
function LazyImage({
  src,
  alt,
  className,
  style,
  zoomable = false,
  priority = false,
  height,
  ...rest
}) {
  var _a, _b, _c;
  const { recordMap, zoom, previewImages, forceCustomImages, components } = useNotionContext();
  const zoomRef = React.useRef(zoom ? zoom.clone() : null);
  const previewImage = previewImages ? (_c = (_a = recordMap == null ? void 0 : recordMap.preview_images) == null ? void 0 : _a[src]) != null ? _c : (_b = recordMap == null ? void 0 : recordMap.preview_images) == null ? void 0 : _b[normalizeUrl(src)] : null;
  const onLoad = React.useCallback(
    (e) => {
      if (zoomable && (e.target.src || e.target.srcset)) {
        if (zoomRef.current) {
          ;
          zoomRef.current.attach(e.target);
        }
      }
    },
    [zoomRef, zoomable]
  );
  const attachZoom = React.useCallback(
    (image) => {
      if (zoomRef.current && image) {
        ;
        zoomRef.current.attach(image);
      }
    },
    [zoomRef]
  );
  const attachZoomRef = React.useMemo(
    () => zoomable ? attachZoom : void 0,
    [zoomable, attachZoom]
  );
  if (previewImage) {
    const aspectRatio = previewImage.originalHeight / previewImage.originalWidth;
    if (components.Image) {
      return /* @__PURE__ */ jsx2(
        components.Image,
        {
          src,
          alt,
          style,
          className,
          width: previewImage.originalWidth,
          height: previewImage.originalHeight,
          blurDataURL: previewImage.dataURIBase64,
          placeholder: "blur",
          priority,
          onLoad
        }
      );
    }
    return /* @__PURE__ */ jsx2(LazyImageFull, { src, ...rest, experimentalDecode: true, children: ({ imageState, ref }) => {
      const isLoaded = imageState === "LoadSuccess" /* LoadSuccess */;
      const wrapperStyle = {
        width: "100%"
      };
      const imgStyle = {};
      if (height) {
        wrapperStyle.height = height;
      } else {
        imgStyle.position = "absolute";
        wrapperStyle.paddingBottom = `${aspectRatio * 100}%`;
      }
      return /* @__PURE__ */ jsxs(
        "div",
        {
          className: cs(
            "lazy-image-wrapper",
            isLoaded && "lazy-image-loaded",
            className
          ),
          style: wrapperStyle,
          children: [
            /* @__PURE__ */ jsx2(
              "img",
              {
                className: "lazy-image-preview",
                src: previewImage.dataURIBase64,
                alt,
                ref,
                style,
                decoding: "async"
              }
            ),
            /* @__PURE__ */ jsx2(
              "img",
              {
                className: "lazy-image-real",
                src,
                alt,
                ref: attachZoomRef,
                style: {
                  ...style,
                  ...imgStyle
                },
                width: previewImage.originalWidth,
                height: previewImage.originalHeight,
                decoding: "async",
                loading: "lazy"
              }
            )
          ]
        }
      );
    } });
  } else {
    if (components.Image && forceCustomImages) {
      return /* @__PURE__ */ jsx2(
        components.Image,
        {
          src,
          alt,
          className,
          style,
          width: null,
          height: height || null,
          priority,
          onLoad
        }
      );
    }
    return /* @__PURE__ */ jsx2(
      "img",
      {
        className,
        style,
        src,
        alt,
        ref: attachZoomRef,
        loading: "lazy",
        decoding: "async",
        ...rest
      }
    );
  }
}

// src/components/lite-youtube-embed.tsx
import React2 from "react";
import { Fragment, jsx as jsx3, jsxs as jsxs2 } from "react/jsx-runtime";
var qs = (params) => {
  return Object.keys(params).map(
    (key) => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`
  ).join("&");
};
var resolutions = [120, 320, 480, 640, 1280];
var resolutionMap = {
  120: "default",
  320: "mqdefault",
  480: "hqdefault",
  640: "sddefault",
  1280: "maxresdefault"
  // 2k, 4k, 8k images don't seem to be available
  // Source: https://longzero.com/articles/youtube-thumbnail-sizes-url/
};
var resolutionSizes = resolutions.map((resolution) => `(max-width: ${resolution}px) ${resolution}px`).join(", ");
function getPosterUrl(id, resolution = 480, type = "jpg") {
  if (type === "webp") {
    return `https://i.ytimg.com/vi_webp/${id}/${resolutionMap[resolution]}.webp`;
  }
  return `https://i.ytimg.com/vi/${id}/${resolutionMap[resolution]}.jpg`;
}
function generateSrcSet(id, type = "jpg") {
  return resolutions.map((resolution) => `${getPosterUrl(id, resolution, type)} ${resolution}w`).join(", ");
}
function LiteYouTubeEmbed({
  id,
  defaultPlay = false,
  mute = false,
  lazyImage = false,
  iframeTitle = "YouTube video",
  alt = "Video preview",
  params = {},
  adLinksPreconnect = true,
  style,
  className
}) {
  const muteParam = mute || defaultPlay ? "1" : "0";
  const queryString = React2.useMemo(
    () => qs({ autoplay: "1", mute: muteParam, ...params }),
    [muteParam, params]
  );
  const ytUrl = "https://www.youtube-nocookie.com";
  const iframeSrc = `${ytUrl}/embed/${id}?${queryString}`;
  const [isPreconnected, setIsPreconnected] = React2.useState(false);
  const [iframeInitialized, setIframeInitialized] = React2.useState(defaultPlay);
  const [isIframeLoaded, setIsIframeLoaded] = React2.useState(false);
  const warmConnections = React2.useCallback(() => {
    if (isPreconnected) return;
    setIsPreconnected(true);
  }, [isPreconnected]);
  const onLoadIframe = React2.useCallback(() => {
    if (iframeInitialized) return;
    setIframeInitialized(true);
  }, [iframeInitialized]);
  const onIframeLoaded = React2.useCallback(() => {
    setIsIframeLoaded(true);
  }, []);
  return /* @__PURE__ */ jsxs2(Fragment, { children: [
    /* @__PURE__ */ jsx3(
      "link",
      {
        rel: "preload",
        as: "image",
        href: getPosterUrl(id),
        imageSrcSet: generateSrcSet(id, "webp"),
        imageSizes: resolutionSizes
      }
    ),
    isPreconnected && /* @__PURE__ */ jsxs2(Fragment, { children: [
      /* @__PURE__ */ jsx3("link", { rel: "preconnect", href: ytUrl }),
      /* @__PURE__ */ jsx3("link", { rel: "preconnect", href: "https://www.google.com" })
    ] }),
    isPreconnected && adLinksPreconnect && /* @__PURE__ */ jsxs2(Fragment, { children: [
      /* @__PURE__ */ jsx3("link", { rel: "preconnect", href: "https://static.doubleclick.net" }),
      /* @__PURE__ */ jsx3("link", { rel: "preconnect", href: "https://googleads.g.doubleclick.net" })
    ] }),
    /* @__PURE__ */ jsxs2(
      "div",
      {
        onClick: onLoadIframe,
        onPointerOver: warmConnections,
        className: cs(
          "notion-yt-lite",
          isIframeLoaded && "notion-yt-loaded",
          iframeInitialized && "notion-yt-initialized",
          className
        ),
        style,
        children: [
          /* @__PURE__ */ jsxs2("picture", { children: [
            resolutions.map((resolution) => /* @__PURE__ */ jsx3(
              "source",
              {
                srcSet: `${getPosterUrl(id, resolution, "webp")} ${resolution}w`,
                media: `(max-width: ${resolution}px)`,
                type: "image/webp"
              },
              resolution
            )),
            /* @__PURE__ */ jsx3(
              "img",
              {
                src: getPosterUrl(id),
                className: "notion-yt-thumbnail",
                loading: lazyImage ? "lazy" : void 0,
                alt
              }
            )
          ] }),
          /* @__PURE__ */ jsx3("div", { className: "notion-yt-playbtn" }),
          iframeInitialized && /* @__PURE__ */ jsx3(
            "iframe",
            {
              width: "560",
              height: "315",
              frameBorder: "0",
              allow: "accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture",
              allowFullScreen: true,
              title: iframeTitle,
              src: iframeSrc,
              onLoad: onIframeLoaded
            }
          )
        ]
      }
    )
  ] });
}

// src/components/asset.tsx
import { Fragment as Fragment2, jsx as jsx4, jsxs as jsxs3 } from "react/jsx-runtime";
var isServer = !globalThis.window;
var supportedAssetTypes = /* @__PURE__ */ new Set([
  "replit",
  "video",
  "image",
  "embed",
  "figma",
  "typeform",
  "excalidraw",
  "maps",
  "tweet",
  "pdf",
  "gist",
  "codepen",
  "drive"
]);
function Asset({
  block,
  zoomable = true,
  children
}) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m;
  const { recordMap, mapImageUrl, components } = useNotionContext();
  if (!block || !supportedAssetTypes.has(block.type)) {
    return null;
  }
  const style = {
    position: "relative",
    display: "flex",
    justifyContent: "center",
    alignSelf: "center",
    width: "100%",
    maxWidth: "100%",
    flexDirection: "column"
  };
  const assetStyle = {};
  if (block.format) {
    const {
      block_aspect_ratio,
      block_height,
      block_width,
      block_full_width,
      block_page_width,
      block_preserve_scale
    } = block.format;
    if (block_full_width || block_page_width) {
      if (block_full_width) {
        style.width = "100vw";
      } else {
        style.width = "100%";
      }
      if (block.type === "video") {
        if (block_height) {
          style.height = block_height;
        } else if (block_aspect_ratio) {
          style.paddingBottom = `${block_aspect_ratio * 100}%`;
        } else if (block_preserve_scale) {
          style.objectFit = "contain";
        }
      } else if (block_aspect_ratio && block.type !== "image") {
        style.paddingBottom = `${block_aspect_ratio * 100}%`;
      } else if (block_height) {
        style.height = block_height;
      } else if (block_preserve_scale) {
        if (block.type === "image") {
          style.height = "100%";
        } else {
          style.paddingBottom = "75%";
          style.minHeight = 100;
        }
      }
    } else {
      switch ((_a = block.format) == null ? void 0 : _a.block_alignment) {
        case "center":
          style.alignSelf = "center";
          break;
        case "left":
          style.alignSelf = "start";
          break;
        case "right":
          style.alignSelf = "end";
          break;
      }
      if (block_width) {
        style.width = block_width;
      }
      if (block_preserve_scale && block.type !== "image") {
        style.paddingBottom = "50%";
        style.minHeight = 100;
      } else {
        if (block_height && block.type !== "image") {
          style.height = block_height;
        }
      }
    }
    if (block.type === "image") {
      assetStyle.objectFit = "cover";
    } else if (block_preserve_scale) {
      assetStyle.objectFit = "contain";
    }
  }
  let source = ((_b = recordMap.signed_urls) == null ? void 0 : _b[block.id]) || ((_e = (_d = (_c = block.properties) == null ? void 0 : _c.source) == null ? void 0 : _d[0]) == null ? void 0 : _e[0]);
  if (!source) {
    return null;
  }
  if (block.space_id) {
    const url = new URL(source);
    url.searchParams.set("spaceId", block.space_id);
    source = url.toString();
  }
  let content = null;
  if (block.type === "tweet") {
    const src = source;
    if (!src) return null;
    const id = (_g = (_f = src.split("?")) == null ? void 0 : _f[0]) == null ? void 0 : _g.split("/").pop();
    if (!id) return null;
    content = /* @__PURE__ */ jsx4(
      "div",
      {
        style: {
          ...assetStyle,
          maxWidth: 420,
          width: "100%",
          marginLeft: "auto",
          marginRight: "auto"
        },
        children: /* @__PURE__ */ jsx4(components.Tweet, { id })
      }
    );
  } else if (block.type === "pdf") {
    style.overflow = "auto";
    style.background = "rgb(226, 226, 226)";
    style.display = "block";
    if (!style.padding) {
      style.padding = "8px 16px";
    }
    if (!isServer) {
      content = /* @__PURE__ */ jsx4(components.Pdf, { file: source });
    }
  } else if (block.type === "embed" || block.type === "video" || block.type === "figma" || block.type === "typeform" || block.type === "gist" || block.type === "maps" || block.type === "excalidraw" || block.type === "codepen" || block.type === "drive" || block.type === "replit") {
    if (block.type === "video" && source && !source.includes("youtube") && !source.includes("youtu.be") && !source.includes("vimeo") && !source.includes("wistia") && !source.includes("loom") && !source.includes("videoask") && !source.includes("getcloudapp") && !source.includes("tella")) {
      style.paddingBottom = void 0;
      content = /* @__PURE__ */ jsx4(
        "video",
        {
          playsInline: true,
          controls: true,
          preload: "metadata",
          style: assetStyle,
          src: source,
          title: block.type
        }
      );
    } else {
      let src = ((_h = block.format) == null ? void 0 : _h.display_source) || source;
      if (src) {
        const youtubeVideoId = block.type === "video" ? getYoutubeId(src) : null;
        if (youtubeVideoId) {
          const params = getUrlParams(src);
          content = /* @__PURE__ */ jsx4(
            LiteYouTubeEmbed,
            {
              id: youtubeVideoId,
              style: assetStyle,
              className: "notion-asset-object-fit",
              params
            }
          );
        } else if (block.type === "gist") {
          if (!src.endsWith(".pibb")) {
            src = `${src}.pibb`;
          }
          assetStyle.width = "100%";
          style.paddingBottom = "50%";
          content = /* @__PURE__ */ jsx4(
            "iframe",
            {
              style: assetStyle,
              className: "notion-asset-object-fit",
              src,
              title: "GitHub Gist",
              frameBorder: "0",
              loading: "lazy",
              scrolling: "auto"
            }
          );
        } else {
          src += block.type === "typeform" ? "&disable-auto-focus=true" : "";
          content = /* @__PURE__ */ jsx4(
            "iframe",
            {
              className: "notion-asset-object-fit",
              style: assetStyle,
              src,
              title: `iframe ${block.type}`,
              frameBorder: "0",
              allowFullScreen: true,
              loading: "lazy",
              scrolling: "auto"
            }
          );
        }
      }
    }
  } else if (block.type === "image") {
    if (!source.includes(".gif") && source.includes("file.notion.so")) {
      source = (_k = (_j = (_i = block.properties) == null ? void 0 : _i.source) == null ? void 0 : _j[0]) == null ? void 0 : _k[0];
    }
    const src = mapImageUrl(source, block);
    const altText = getTextContent((_l = block.properties) == null ? void 0 : _l.alt_text);
    const caption = getTextContent((_m = block.properties) == null ? void 0 : _m.caption);
    const alt = altText || caption || "notion image";
    content = /* @__PURE__ */ jsx4(
      LazyImage,
      {
        src,
        alt,
        zoomable,
        height: style.height,
        style: assetStyle
      }
    );
  }
  return /* @__PURE__ */ jsxs3(Fragment2, { children: [
    /* @__PURE__ */ jsxs3("div", { style, children: [
      content,
      block.type === "image" && children
    ] }),
    block.type !== "image" && children
  ] });
}

// src/components/text.tsx
import "notion-types";
import { getBlockValue, parsePageId } from "notion-utils";
import React8 from "react";

// src/components/eoi.tsx
import "notion-types";

// src/icons/type-github.tsx
import { jsx as jsx5 } from "react/jsx-runtime";
function SvgTypeGitHub(props) {
  return /* @__PURE__ */ jsx5("svg", { viewBox: "0 0 260 260", ...props, children: /* @__PURE__ */ jsx5("g", { children: /* @__PURE__ */ jsx5(
    "path",
    {
      d: "M128.00106,0 C57.3172926,0 0,57.3066942 0,128.00106 C0,184.555281 36.6761997,232.535542 87.534937,249.460899 C93.9320223,250.645779 96.280588,246.684165 96.280588,243.303333 C96.280588,240.251045 96.1618878,230.167899 96.106777,219.472176 C60.4967585,227.215235 52.9826207,204.369712 52.9826207,204.369712 C47.1599584,189.574598 38.770408,185.640538 38.770408,185.640538 C27.1568785,177.696113 39.6458206,177.859325 39.6458206,177.859325 C52.4993419,178.762293 59.267365,191.04987 59.267365,191.04987 C70.6837675,210.618423 89.2115753,204.961093 96.5158685,201.690482 C97.6647155,193.417512 100.981959,187.77078 104.642583,184.574357 C76.211799,181.33766 46.324819,170.362144 46.324819,121.315702 C46.324819,107.340889 51.3250588,95.9223682 59.5132437,86.9583937 C58.1842268,83.7344152 53.8029229,70.715562 60.7532354,53.0843636 C60.7532354,53.0843636 71.5019501,49.6441813 95.9626412,66.2049595 C106.172967,63.368876 117.123047,61.9465949 128.00106,61.8978432 C138.879073,61.9465949 149.837632,63.368876 160.067033,66.2049595 C184.49805,49.6441813 195.231926,53.0843636 195.231926,53.0843636 C202.199197,70.715562 197.815773,83.7344152 196.486756,86.9583937 C204.694018,95.9223682 209.660343,107.340889 209.660343,121.315702 C209.660343,170.478725 179.716133,181.303747 151.213281,184.472614 C155.80443,188.444828 159.895342,196.234518 159.895342,208.176593 C159.895342,225.303317 159.746968,239.087361 159.746968,243.303333 C159.746968,246.709601 162.05102,250.70089 168.53925,249.443941 C219.370432,232.499507 256,184.536204 256,128.00106 C256,57.3066942 198.691187,0 128.00106,0 Z M47.9405593,182.340212 C47.6586465,182.976105 46.6581745,183.166873 45.7467277,182.730227 C44.8183235,182.312656 44.2968914,181.445722 44.5978808,180.80771 C44.8734344,180.152739 45.876026,179.97045 46.8023103,180.409216 C47.7328342,180.826786 48.2627451,181.702199 47.9405593,182.340212 Z M54.2367892,187.958254 C53.6263318,188.524199 52.4329723,188.261363 51.6232682,187.366874 C50.7860088,186.474504 50.6291553,185.281144 51.2480912,184.70672 C51.8776254,184.140775 53.0349512,184.405731 53.8743302,185.298101 C54.7115892,186.201069 54.8748019,187.38595 54.2367892,187.958254 Z M58.5562413,195.146347 C57.7719732,195.691096 56.4895886,195.180261 55.6968417,194.042013 C54.9125733,192.903764 54.9125733,191.538713 55.713799,190.991845 C56.5086651,190.444977 57.7719732,190.936735 58.5753181,192.066505 C59.3574669,193.22383 59.3574669,194.58888 58.5562413,195.146347 Z M65.8613592,203.471174 C65.1597571,204.244846 63.6654083,204.03712 62.5716717,202.981538 C61.4524999,201.94927 61.1409122,200.484596 61.8446341,199.710926 C62.5547146,198.935137 64.0575422,199.15346 65.1597571,200.200564 C66.2704506,201.230712 66.6095936,202.705984 65.8613592,203.471174 Z M75.3025151,206.281542 C74.9930474,207.284134 73.553809,207.739857 72.1039724,207.313809 C70.6562556,206.875043 69.7087748,205.700761 70.0012857,204.687571 C70.302275,203.678621 71.7478721,203.20382 73.2083069,203.659543 C74.6539041,204.09619 75.6035048,205.261994 75.3025151,206.281542 Z M86.046947,207.473627 C86.0829806,208.529209 84.8535871,209.404622 83.3316829,209.4237 C81.8013,209.457614 80.563428,208.603398 80.5464708,207.564772 C80.5464708,206.498591 81.7483088,205.631657 83.2786917,205.606221 C84.8005962,205.576546 86.046947,206.424403 86.046947,207.473627 Z M96.6021471,207.069023 C96.7844366,208.099171 95.7267341,209.156872 94.215428,209.438785 C92.7295577,209.710099 91.3539086,209.074206 91.1652603,208.052538 C90.9808515,206.996955 92.0576306,205.939253 93.5413813,205.66582 C95.054807,205.402984 96.4092596,206.021919 96.6021471,207.069023 Z",
      fill: "#161614"
    }
  ) }) });
}
var type_github_default = SvgTypeGitHub;

// src/components/mention-preview-card.tsx
import { jsx as jsx6, jsxs as jsxs4 } from "react/jsx-runtime";
function capitalizeFirstLetter(str) {
  if (!str) return "";
  return str.charAt(0).toUpperCase() + str.slice(1);
}
function MentionPreviewCard({
  owner,
  lastUpdated,
  externalImage,
  title,
  domain
}) {
  return /* @__PURE__ */ jsxs4("div", { className: "notion-external-subtitle", children: [
    externalImage && /* @__PURE__ */ jsxs4("div", { className: "notion-preview-card-domain-warp", children: [
      /* @__PURE__ */ jsx6("div", { className: "notion-preview-card-logo", children: externalImage }),
      /* @__PURE__ */ jsx6("div", { className: "notion-preview-card-domain", children: capitalizeFirstLetter(domain.split(".")[0]) })
    ] }),
    /* @__PURE__ */ jsx6("div", { className: "notion-preview-card-title", children: title }),
    owner && /* @__PURE__ */ jsxs4("div", { className: "notion-external-subtitle-item", children: [
      /* @__PURE__ */ jsx6("div", { className: "notion-external-subtitle-item-name", children: "Owner" }),
      /* @__PURE__ */ jsx6("span", { className: "notion-external-subtitle-item-desc", children: owner })
    ] }),
    lastUpdated && /* @__PURE__ */ jsxs4("div", { className: "notion-external-subtitle-item", children: [
      /* @__PURE__ */ jsx6("div", { className: "notion-external-subtitle-item-name", children: "Updated" }),
      /* @__PURE__ */ jsx6("span", { className: "notion-external-subtitle-item-desc", children: lastUpdated })
    ] }),
    domain === "github.com" && /* @__PURE__ */ jsxs4("div", { className: "notion-preview-card-github-shields", children: [
      /* @__PURE__ */ jsx6(
        "img",
        {
          src: `https://img.shields.io/github/stars/${owner}/${title}?logo=github`,
          alt: ""
        }
      ),
      /* @__PURE__ */ jsx6(
        "img",
        {
          src: `https://img.shields.io/github/last-commit/${owner}/${title}`,
          alt: ""
        }
      )
    ] })
  ] });
}

// src/components/eoi.tsx
import { jsx as jsx7, jsxs as jsxs5 } from "react/jsx-runtime";
function EOI({
  block,
  inline,
  className
}) {
  var _a, _b, _c;
  const { components } = useNotionContext();
  const { original_url, attributes, domain } = (block == null ? void 0 : block.format) || {};
  if (!original_url || !attributes) {
    return null;
  }
  const title = (_a = attributes.find((attr) => attr.id === "title")) == null ? void 0 : _a.values[0];
  let owner = (_b = attributes.find((attr) => attr.id === "owner")) == null ? void 0 : _b.values[0];
  const lastUpdatedAt = (_c = attributes.find((attr) => attr.id === "updated_at")) == null ? void 0 : _c.values[0];
  const lastUpdated = lastUpdatedAt ? formatNotionDateTime(lastUpdatedAt) : null;
  let externalImage;
  switch (domain) {
    case "github.com":
      externalImage = /* @__PURE__ */ jsx7(type_github_default, {});
      if (owner) {
        const parts = owner.split("/");
        owner = parts.at(-1);
      }
      break;
    default:
      if (true) {
        console.log(
          `Unsupported external_object_instance domain "${domain}"`,
          JSON.stringify(block, null, 2)
        );
      }
      return null;
  }
  return /* @__PURE__ */ jsxs5(
    components.Link,
    {
      target: "_blank",
      rel: "noopener noreferrer",
      href: original_url,
      className: cs(
        "notion-external",
        inline ? "notion-external-mention" : "notion-external-block notion-row",
        className
      ),
      children: [
        externalImage && /* @__PURE__ */ jsx7("div", { className: "notion-external-image", children: externalImage }),
        /* @__PURE__ */ jsxs5("div", { className: "notion-external-description", children: [
          /* @__PURE__ */ jsx7("div", { className: "notion-external-title", children: title }),
          !inline && owner ? /* @__PURE__ */ jsxs5("div", { className: "notion-external-block-desc", children: [
            owner,
            lastUpdated && /* @__PURE__ */ jsx7("span", { children: " \u2022 " }),
            lastUpdated && `Updated ${lastUpdated}`
          ] }) : null,
          inline && (owner || lastUpdated) && /* @__PURE__ */ jsx7(
            MentionPreviewCard,
            {
              title,
              owner,
              lastUpdated,
              domain,
              externalImage
            }
          )
        ] })
      ]
    }
  );
}

// src/components/graceful-image.tsx
import "react";
import { Img } from "react-image";
import { jsx as jsx8 } from "react/jsx-runtime";
function GracefulImage(props) {
  if (isBrowser) {
    return /* @__PURE__ */ jsx8(Img, { ...props });
  } else {
    return /* @__PURE__ */ jsx8("img", { ...props });
  }
}

// src/components/link-mention.tsx
import "react";
import { jsx as jsx9, jsxs as jsxs6 } from "react/jsx-runtime";
function LinkMention({ metadata }) {
  return /* @__PURE__ */ jsxs6("span", { className: "notion-link-mention", children: [
    /* @__PURE__ */ jsx9(LinkMentionInline, { metadata }),
    /* @__PURE__ */ jsx9(LinkMentionPreview, { metadata })
  ] });
}
function LinkMentionInline({ metadata }) {
  return /* @__PURE__ */ jsxs6(
    "a",
    {
      href: metadata.href,
      target: "_blank",
      rel: "noopener noreferrer",
      className: "notion-link-mention-link",
      children: [
        /* @__PURE__ */ jsx9(
          "img",
          {
            className: "notion-link-mention-icon",
            src: metadata.icon_url,
            alt: metadata.link_provider
          }
        ),
        metadata.link_provider && /* @__PURE__ */ jsx9("span", { className: "notion-link-mention-provider", children: metadata.link_provider }),
        /* @__PURE__ */ jsx9("span", { className: "notion-link-mention-title", children: metadata.title })
      ]
    }
  );
}
function LinkMentionPreview({ metadata }) {
  return /* @__PURE__ */ jsx9("div", { className: "notion-link-mention-preview", children: /* @__PURE__ */ jsxs6("article", { className: "notion-link-mention-card", children: [
    /* @__PURE__ */ jsx9(
      "img",
      {
        className: "notion-link-mention-preview-thumbnail",
        src: metadata.thumbnail_url,
        alt: metadata.title,
        referrerPolicy: "same-origin"
      }
    ),
    /* @__PURE__ */ jsxs6("div", { className: "notion-link-mention-preview-content", children: [
      /* @__PURE__ */ jsx9("p", { className: "notion-link-mention-preview-title", children: metadata.title }),
      /* @__PURE__ */ jsx9("p", { className: "notion-link-mention-preview-description", children: metadata.description }),
      /* @__PURE__ */ jsxs6("div", { className: "notion-link-mention-preview-footer", children: [
        /* @__PURE__ */ jsx9(
          "img",
          {
            className: "notion-link-mention-preview-icon",
            src: metadata.icon_url,
            alt: metadata.link_provider,
            referrerPolicy: "same-origin"
          }
        ),
        /* @__PURE__ */ jsx9("span", { className: "notion-link-mention-preview-provider", children: metadata.link_provider })
      ] })
    ] })
  ] }) });
}

// src/components/page-title.tsx
import "notion-types";
import { getBlockTitle as getBlockTitle2 } from "notion-utils";
import React7 from "react";

// src/components/page-icon.tsx
import "notion-types";
import { getBlockIcon, getBlockTitle } from "notion-utils";
import React6 from "react";

// src/icons/default-page-icon.tsx
import "react";
import { jsx as jsx10 } from "react/jsx-runtime";
function DefaultPageIcon(props) {
  const { className, ...rest } = props;
  return /* @__PURE__ */ jsx10("svg", { className, ...rest, viewBox: "0 0 30 30", width: "16", children: /* @__PURE__ */ jsx10("path", { d: "M16,1H4v28h22V11L16,1z M16,3.828L23.172,11H16V3.828z M24,27H6V3h8v10h10V27z M8,17h14v-2H8V17z M8,21h14v-2H8V21z M8,25h14v-2H8V25z" }) });
}

// src/components/page-icon.tsx
import { jsx as jsx11 } from "react/jsx-runtime";
var isIconBlock = (value) => {
  return value.type === "page" || value.type === "callout" || value.type === "collection_view" || value.type === "collection_view_page";
};
function PageIconImpl({
  block,
  className,
  inline = true,
  hideDefaultIcon = false,
  defaultIcon
}) {
  var _a;
  const { mapImageUrl, recordMap, darkMode } = useNotionContext();
  let isImage = false;
  let content = null;
  if (isIconBlock(block)) {
    const icon = ((_a = getBlockIcon(block, recordMap)) == null ? void 0 : _a.trim()) || defaultIcon;
    const title = getBlockTitle(block, recordMap);
    if (icon && isUrl(icon)) {
      const url = mapImageUrl(icon, block);
      isImage = true;
      content = /* @__PURE__ */ jsx11(
        LazyImage,
        {
          src: url,
          alt: title || "page icon",
          className: cs(className, "notion-page-icon")
        }
      );
    } else if (icon && icon.startsWith("/icons/")) {
      const url = "https://www.notion.so" + icon + "?mode=" + (darkMode ? "dark" : "light");
      content = /* @__PURE__ */ jsx11(
        LazyImage,
        {
          src: url,
          alt: title || "page icon",
          className: cs(className, "notion-page-icon")
        }
      );
    } else if (!icon) {
      if (!hideDefaultIcon) {
        isImage = true;
        content = /* @__PURE__ */ jsx11(
          DefaultPageIcon,
          {
            className: cs(className, "notion-page-icon"),
            alt: title || "page icon"
          }
        );
      }
    } else {
      isImage = false;
      content = /* @__PURE__ */ jsx11(
        "span",
        {
          className: cs(className, "notion-page-icon"),
          role: "img",
          "aria-label": icon,
          children: icon
        }
      );
    }
  }
  if (!content) {
    return null;
  }
  return /* @__PURE__ */ jsx11(
    "div",
    {
      className: cs(
        inline ? "notion-page-icon-inline" : "notion-page-icon-hero",
        isImage ? "notion-page-icon-image" : "notion-page-icon-span"
      ),
      children: content
    }
  );
}
var PageIcon = React6.memo(PageIconImpl);

// src/components/page-title.tsx
import { jsx as jsx12, jsxs as jsxs7 } from "react/jsx-runtime";
function PageTitleImpl({
  block,
  className,
  defaultIcon,
  ...rest
}) {
  var _a, _b;
  const { recordMap } = useNotionContext();
  if (!block) return null;
  if (block.type === "collection_view_page" || block.type === "collection_view") {
    const title = getBlockTitle2(block, recordMap);
    if (!title) {
      return null;
    }
    const titleDecoration = [[title]];
    return /* @__PURE__ */ jsxs7("span", { className: cs("notion-page-title", className), ...rest, children: [
      /* @__PURE__ */ jsx12(
        PageIcon,
        {
          block,
          defaultIcon,
          className: "notion-page-title-icon"
        }
      ),
      /* @__PURE__ */ jsx12("span", { className: "notion-page-title-text", children: /* @__PURE__ */ jsx12(Text, { value: titleDecoration, block }) })
    ] });
  }
  if (!((_a = block.properties) == null ? void 0 : _a.title)) {
    return null;
  }
  return /* @__PURE__ */ jsxs7("span", { className: cs("notion-page-title", className), ...rest, children: [
    /* @__PURE__ */ jsx12(
      PageIcon,
      {
        block,
        defaultIcon,
        className: "notion-page-title-icon"
      }
    ),
    /* @__PURE__ */ jsx12("span", { className: "notion-page-title-text", children: /* @__PURE__ */ jsx12(Text, { value: (_b = block.properties) == null ? void 0 : _b.title, block }) })
  ] });
}
var PageTitle = React7.memo(PageTitleImpl);

// src/components/text.tsx
import { Fragment as Fragment3, jsx as jsx13 } from "react/jsx-runtime";
function Text({
  value,
  block,
  linkProps,
  linkProtocol
}) {
  const { components, recordMap, mapPageUrl, mapImageUrl, rootDomain } = useNotionContext();
  return /* @__PURE__ */ jsx13(React8.Fragment, { children: value == null ? void 0 : value.map(([text, decorations], index) => {
    if (!decorations) {
      if (text === ",") {
        return /* @__PURE__ */ jsx13("span", { style: { padding: "0.5em" } }, index);
      } else {
        return /* @__PURE__ */ jsx13(React8.Fragment, { children: text }, index);
      }
    }
    const formatted = decorations.reduce(
      (element, decorator) => {
        switch (decorator[0]) {
          case "p": {
            const blockId = decorator[1];
            const linkedBlock = getBlockValue(recordMap.block[blockId]);
            if (!linkedBlock) {
              console.log('"p" missing block', blockId);
              return null;
            }
            return /* @__PURE__ */ jsx13(
              components.PageLink,
              {
                className: "notion-link",
                href: mapPageUrl(blockId),
                children: /* @__PURE__ */ jsx13(PageTitle, { block: linkedBlock })
              }
            );
          }
          case "\u2023": {
            const linkType = decorator[1][0];
            const id = decorator[1][1];
            switch (linkType) {
              case "u": {
                const user = getBlockValue(recordMap.notion_user[id]);
                if (!user) {
                  console.log('"\u2023" missing user', id);
                  return null;
                }
                const src = mapImageUrl(user.profile_photo, block);
                if (!src) return null;
                const name = [user.given_name, user.family_name].filter(Boolean).join(" ");
                return /* @__PURE__ */ jsx13(
                  GracefulImage,
                  {
                    className: "notion-user",
                    src,
                    alt: name
                  }
                );
              }
              default: {
                const linkedBlock = getBlockValue(recordMap.block[id]);
                if (!linkedBlock) {
                  console.log('"\u2023" missing block', linkType, id);
                  return null;
                }
                return /* @__PURE__ */ jsx13(
                  components.PageLink,
                  {
                    className: "notion-link",
                    href: mapPageUrl(id),
                    ...linkProps,
                    target: "_blank",
                    rel: "noopener noreferrer",
                    children: /* @__PURE__ */ jsx13(PageTitle, { block: linkedBlock })
                  }
                );
              }
            }
          }
          case "h":
            return /* @__PURE__ */ jsx13("span", { className: `notion-${decorator[1]}`, children: element });
          case "c":
            return /* @__PURE__ */ jsx13("code", { className: "notion-inline-code", children: element });
          case "b":
            return /* @__PURE__ */ jsx13("b", { children: element });
          case "i":
            return /* @__PURE__ */ jsx13("em", { children: element });
          case "s":
            return /* @__PURE__ */ jsx13("s", { children: element });
          case "_":
            return /* @__PURE__ */ jsx13("span", { className: "notion-inline-underscore", children: element });
          case "e":
            return /* @__PURE__ */ jsx13(components.Equation, { math: decorator[1], inline: true });
          case "m":
            return element;
          //still need to return the base element
          case "a": {
            const v = decorator[1];
            const pathname = v.slice(1);
            const id = parsePageId(pathname, { uuid: true });
            if (rootDomain && v.includes(rootDomain) || id && v[0] === "/") {
              const href = rootDomain && v.includes(rootDomain) ? v : `${mapPageUrl(id)}${getHashFragmentValue(v)}`;
              return /* @__PURE__ */ jsx13(
                components.PageLink,
                {
                  className: "notion-link",
                  href,
                  ...linkProps,
                  children: element
                }
              );
            } else {
              return /* @__PURE__ */ jsx13(
                components.Link,
                {
                  className: "notion-link",
                  href: linkProtocol ? `${linkProtocol}:${decorator[1]}` : decorator[1],
                  ...linkProps,
                  children: element
                }
              );
            }
          }
          case "d": {
            const v = decorator[1];
            const type = v == null ? void 0 : v.type;
            if (type === "date") {
              const startDate = v.start_date;
              return formatDate(startDate);
            } else if (type === "datetime") {
              const startDate = v.start_date;
              const startTime = v.start_time;
              return `${formatDate(startDate)} ${startTime}`;
            } else if (type === "daterange") {
              const startDate = v.start_date;
              const endDate = v.end_date;
              return `${formatDate(startDate)} \u2192 ${formatDate(endDate)}`;
            } else {
              return element;
            }
          }
          case "u": {
            const userId = decorator[1];
            const user = getBlockValue(recordMap.notion_user[userId]);
            if (!user) {
              console.log("missing user", userId);
              return null;
            }
            const src = mapImageUrl(user.profile_photo, block);
            if (!src) return null;
            const name = [user.given_name, user.family_name].filter(Boolean).join(" ");
            return /* @__PURE__ */ jsx13(GracefulImage, { className: "notion-user", src, alt: name });
          }
          case "lm": {
            const metadata = decorator[1];
            return /* @__PURE__ */ jsx13(LinkMention, { metadata });
          }
          case "eoi": {
            const blockId = decorator[1];
            const externalObjectInstance = getBlockValue(
              recordMap.block[blockId]
            );
            if (!externalObjectInstance) {
              console.log('"eoi" missing block', blockId);
              return null;
            }
            return /* @__PURE__ */ jsx13(EOI, { block: externalObjectInstance, inline: true });
          }
          case "si":
            return null;
          default:
            if (true) {
              console.log("unsupported text format", decorator);
            }
            return element;
        }
      },
      /* @__PURE__ */ jsx13(Fragment3, { children: text })
    );
    return /* @__PURE__ */ jsx13(React8.Fragment, { children: formatted }, index);
  }) });
}

// src/components/asset-wrapper.tsx
import { jsx as jsx14 } from "react/jsx-runtime";
var urlStyle = { width: "100%" };
function AssetWrapper({
  blockId,
  block
}) {
  var _a, _b, _c, _d, _e, _f, _g, _h;
  const value = block;
  const { components, mapPageUrl, rootDomain, zoom } = useNotionContext();
  let isURL = false;
  if (block.type === "image") {
    const caption = (_c = (_b = (_a = value == null ? void 0 : value.properties) == null ? void 0 : _a.caption) == null ? void 0 : _b[0]) == null ? void 0 : _c[0];
    if (caption) {
      const id = parsePageId2(caption, { uuid: true });
      const isPage = caption.charAt(0) === "/" && id;
      if (isPage || isValidURL(caption)) {
        isURL = true;
      }
    }
  }
  const figure = /* @__PURE__ */ jsx14(
    "figure",
    {
      className: cs(
        "notion-asset-wrapper",
        `notion-asset-wrapper-${block.type}`,
        ((_d = value.format) == null ? void 0 : _d.block_full_width) && "notion-asset-wrapper-full",
        blockId
      ),
      children: /* @__PURE__ */ jsx14(Asset, { block: value, zoomable: zoom && !isURL, children: ((_e = value == null ? void 0 : value.properties) == null ? void 0 : _e.caption) && !isURL && /* @__PURE__ */ jsx14("figcaption", { className: "notion-asset-caption", children: /* @__PURE__ */ jsx14(Text, { value: value.properties.caption, block }) }) })
    }
  );
  if (isURL) {
    const caption = (_h = (_g = (_f = value == null ? void 0 : value.properties) == null ? void 0 : _f.caption) == null ? void 0 : _g[0]) == null ? void 0 : _h[0];
    const id = parsePageId2(caption, { uuid: true });
    const isPage = (caption == null ? void 0 : caption.charAt(0)) === "/" && id;
    const captionHostname = extractHostname(caption);
    return /* @__PURE__ */ jsx14(
      components.PageLink,
      {
        style: urlStyle,
        href: isPage ? mapPageUrl(id) : caption,
        target: captionHostname && captionHostname !== rootDomain && !(caption == null ? void 0 : caption.startsWith("/")) ? "blank_" : null,
        children: figure
      }
    );
  }
  return figure;
}
function isValidURL(str) {
  const pattern = new RegExp(
    "^(https?:\\/\\/)?((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|((\\d{1,3}\\.){3}\\d{1,3}))(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*(\\?[;&a-z\\d%_.~+=-]*)?(\\#[-a-z\\d_]*)?$",
    "i"
  );
  return !!pattern.test(str);
}
function extractHostname(url) {
  try {
    const hostname = new URL(url).hostname;
    return hostname;
  } catch (e) {
    return "";
  }
}

// src/icons/check.tsx
import { jsx as jsx15 } from "react/jsx-runtime";
function SvgCheck(props) {
  return /* @__PURE__ */ jsx15("svg", { viewBox: "0 0 14 14", ...props, children: /* @__PURE__ */ jsx15("path", { d: "M5.5 12L14 3.5 12.5 2l-7 7-4-4.003L0 6.499z" }) });
}
var check_default = SvgCheck;

// src/components/checkbox.tsx
import { jsx as jsx16 } from "react/jsx-runtime";
function Checkbox({
  isChecked
}) {
  let content = null;
  if (isChecked) {
    content = /* @__PURE__ */ jsx16("div", { className: "notion-property-checkbox-checked", children: /* @__PURE__ */ jsx16(check_default, {}) });
  } else {
    content = /* @__PURE__ */ jsx16("div", { className: "notion-property-checkbox-unchecked" });
  }
  return /* @__PURE__ */ jsx16("span", { className: "notion-property notion-property-checkbox", children: content });
}

// src/components/header.tsx
import { getPageBreadcrumbs } from "notion-utils";
import React13 from "react";
import { useHotkeys } from "react-hotkeys-hook";

// src/icons/search-icon.tsx
import "react";
import { jsx as jsx17 } from "react/jsx-runtime";
function SearchIcon(props) {
  const { className, ...rest } = props;
  return /* @__PURE__ */ jsx17("svg", { className: cs("notion-icon", className), viewBox: "0 0 17 17", ...rest, children: /* @__PURE__ */ jsx17("path", { d: "M6.78027 13.6729C8.24805 13.6729 9.60156 13.1982 10.709 12.4072L14.875 16.5732C15.0684 16.7666 15.3232 16.8633 15.5957 16.8633C16.167 16.8633 16.5713 16.4238 16.5713 15.8613C16.5713 15.5977 16.4834 15.3516 16.29 15.1582L12.1504 11.0098C13.0205 9.86719 13.5391 8.45215 13.5391 6.91406C13.5391 3.19629 10.498 0.155273 6.78027 0.155273C3.0625 0.155273 0.0214844 3.19629 0.0214844 6.91406C0.0214844 10.6318 3.0625 13.6729 6.78027 13.6729ZM6.78027 12.2139C3.87988 12.2139 1.48047 9.81445 1.48047 6.91406C1.48047 4.01367 3.87988 1.61426 6.78027 1.61426C9.68066 1.61426 12.0801 4.01367 12.0801 6.91406C12.0801 9.81445 9.68066 12.2139 6.78027 12.2139Z" }) });
}

// src/components/search-dialog.tsx
var import_lodash = __toESM(require_lodash(), 1);
import { getBlockParentPage, getBlockTitle as getBlockTitle3 } from "notion-utils";
import React12 from "react";

// src/icons/clear-icon.tsx
import "react";
import { jsx as jsx18 } from "react/jsx-runtime";
function ClearIcon(props) {
  const { className, ...rest } = props;
  return /* @__PURE__ */ jsx18("svg", { className: cs("notion-icon", className), ...rest, viewBox: "0 0 30 30", children: /* @__PURE__ */ jsx18("path", { d: "M15,0C6.716,0,0,6.716,0,15s6.716,15,15,15s15-6.716,15-15S23.284,0,15,0z M22,20.6L20.6,22L15,16.4L9.4,22L8,20.6l5.6-5.6 L8,9.4L9.4,8l5.6,5.6L20.6,8L22,9.4L16.4,15L22,20.6z" }) });
}

// src/icons/loading-icon.tsx
import "react";
import { jsx as jsx19, jsxs as jsxs8 } from "react/jsx-runtime";
function LoadingIcon(props) {
  const { className, ...rest } = props;
  return /* @__PURE__ */ jsxs8("svg", { className: cs("notion-icon", className), ...rest, viewBox: "0 0 24 24", children: [
    /* @__PURE__ */ jsx19("defs", { children: /* @__PURE__ */ jsxs8(
      "linearGradient",
      {
        x1: "28.1542969%",
        y1: "63.7402344%",
        x2: "74.6289062%",
        y2: "17.7832031%",
        id: "linearGradient-1",
        children: [
          /* @__PURE__ */ jsx19("stop", { stopColor: "rgba(164, 164, 164, 1)", offset: "0%" }),
          /* @__PURE__ */ jsx19(
            "stop",
            {
              stopColor: "rgba(164, 164, 164, 0)",
              stopOpacity: "0",
              offset: "100%"
            }
          )
        ]
      }
    ) }),
    /* @__PURE__ */ jsx19("g", { id: "Page-1", stroke: "none", strokeWidth: "1", fill: "none", children: /* @__PURE__ */ jsx19("g", { transform: "translate(-236.000000, -286.000000)", children: /* @__PURE__ */ jsxs8("g", { transform: "translate(238.000000, 286.000000)", children: [
      /* @__PURE__ */ jsx19(
        "circle",
        {
          id: "Oval-2",
          stroke: "url(#linearGradient-1)",
          strokeWidth: "4",
          cx: "10",
          cy: "12",
          r: "10"
        }
      ),
      /* @__PURE__ */ jsx19(
        "path",
        {
          d: "M10,2 C4.4771525,2 0,6.4771525 0,12",
          id: "Oval-2",
          stroke: "rgba(164, 164, 164, 1)",
          strokeWidth: "4"
        }
      ),
      /* @__PURE__ */ jsx19(
        "rect",
        {
          id: "Rectangle-1",
          fill: "rgba(164, 164, 164, 1)",
          x: "8",
          y: "0",
          width: "4",
          height: "4",
          rx: "8"
        }
      )
    ] }) }) })
  ] });
}

// src/components/search-dialog.tsx
import { Fragment as Fragment4, jsx as jsx20, jsxs as jsxs9 } from "react/jsx-runtime";
var SearchDialog = class extends React12.Component {
  constructor(props) {
    super(props);
    __publicField(this, "state", {
      isLoading: false,
      query: "",
      searchResult: null,
      searchError: null
    });
    __publicField(this, "_inputRef");
    __publicField(this, "_search");
    __publicField(this, "_onAfterOpen", () => {
      if (this._inputRef.current) {
        this._inputRef.current.focus();
      }
    });
    __publicField(this, "_onChangeQuery", (e) => {
      const query = e.target.value;
      this.setState({ query });
      if (!query.trim()) {
        this.setState({ isLoading: false, searchResult: null, searchError: null });
        return;
      } else {
        this._search();
      }
    });
    __publicField(this, "_onClearQuery", () => {
      this._onChangeQuery({ target: { value: "" } });
    });
    __publicField(this, "_warmupSearch", async () => {
      const { searchNotion, rootBlockId } = this.props;
      await searchNotion({
        query: "",
        ancestorId: rootBlockId
      });
    });
    __publicField(this, "_searchImpl", async () => {
      const { searchNotion, rootBlockId } = this.props;
      const { query } = this.state;
      if (!query.trim()) {
        this.setState({ isLoading: false, searchResult: null, searchError: null });
        return;
      }
      this.setState({ isLoading: true });
      const result = await searchNotion({
        query,
        ancestorId: rootBlockId
      });
      console.log("search", query, result);
      let searchResult = null;
      let searchError = null;
      if (result.error || result.errorId) {
        searchError = result;
      } else {
        searchResult = { ...result };
        const results = searchResult.results.map((result2) => {
          var _a, _b;
          const block = (_a = searchResult.recordMap.block[result2.id]) == null ? void 0 : _a.value;
          if (!block) return;
          const title = getBlockTitle3(block, searchResult.recordMap);
          if (!title) {
            return;
          }
          result2.title = title;
          result2.block = block;
          result2.recordMap = searchResult.recordMap;
          result2.page = getBlockParentPage(block, searchResult.recordMap, {
            inclusive: true
          }) || block;
          if (!result2.page.id) {
            return;
          }
          if ((_b = result2.highlight) == null ? void 0 : _b.text) {
            result2.highlight.html = result2.highlight.text.replaceAll(/<gzknfouu>/gi, "<b>").replaceAll(/<\/gzknfouu>/gi, "</b>");
          }
          return result2;
        }).filter(Boolean);
        const searchResultsMap = Object.fromEntries(
          results.map((result2) => [result2.page.id, result2])
        );
        searchResult.results = Object.values(searchResultsMap);
      }
      if (this.state.query === query) {
        this.setState({ isLoading: false, searchResult, searchError });
      }
    });
    this._inputRef = React12.createRef();
  }
  componentDidMount() {
    this._search = (0, import_lodash.default)(this._searchImpl.bind(this), 1e3);
    void this._warmupSearch();
  }
  render() {
    const { isOpen, onClose } = this.props;
    const { isLoading, query, searchResult, searchError } = this.state;
    const hasQuery = !!query.trim();
    return /* @__PURE__ */ jsx20(NotionContextConsumer, { children: (ctx2) => {
      const { components, defaultPageIcon, mapPageUrl } = ctx2;
      return /* @__PURE__ */ jsx20(
        components.Modal,
        {
          isOpen,
          contentLabel: "Search",
          className: "notion-search",
          overlayClassName: "notion-search-overlay",
          onRequestClose: onClose,
          onAfterOpen: this._onAfterOpen,
          children: /* @__PURE__ */ jsxs9("div", { className: "quickFindMenu", children: [
            /* @__PURE__ */ jsxs9("div", { className: "searchBar", children: [
              /* @__PURE__ */ jsx20("div", { className: "inlineIcon", children: isLoading ? /* @__PURE__ */ jsx20(LoadingIcon, { className: "loadingIcon" }) : /* @__PURE__ */ jsx20(SearchIcon, {}) }),
              /* @__PURE__ */ jsx20(
                "input",
                {
                  className: "searchInput",
                  placeholder: "Search",
                  value: query,
                  ref: this._inputRef,
                  onChange: this._onChangeQuery
                }
              ),
              query && /* @__PURE__ */ jsx20(
                "div",
                {
                  role: "button",
                  className: "clearButton",
                  onClick: this._onClearQuery,
                  children: /* @__PURE__ */ jsx20(ClearIcon, { className: "clearIcon" })
                }
              )
            ] }),
            hasQuery && searchResult && /* @__PURE__ */ jsx20(Fragment4, { children: searchResult.results.length ? /* @__PURE__ */ jsxs9(
              NotionContextProvider,
              {
                ...ctx2,
                recordMap: searchResult.recordMap,
                children: [
                  /* @__PURE__ */ jsx20("div", { className: "resultsPane", children: searchResult.results.map((result) => {
                    var _a;
                    return /* @__PURE__ */ jsxs9(
                      components.PageLink,
                      {
                        className: cs("result", "notion-page-link"),
                        href: mapPageUrl(
                          result.page.id,
                          // TODO
                          searchResult.recordMap
                        ),
                        children: [
                          /* @__PURE__ */ jsx20(
                            PageTitle,
                            {
                              block: result.page,
                              defaultIcon: defaultPageIcon
                            }
                          ),
                          ((_a = result.highlight) == null ? void 0 : _a.html) && /* @__PURE__ */ jsx20(
                            "div",
                            {
                              className: "notion-search-result-highlight",
                              dangerouslySetInnerHTML: {
                                __html: result.highlight.html
                              }
                            }
                          )
                        ]
                      },
                      result.id
                    );
                  }) }),
                  /* @__PURE__ */ jsx20("footer", { className: "resultsFooter", children: /* @__PURE__ */ jsxs9("div", { children: [
                    /* @__PURE__ */ jsx20("span", { className: "resultsCount", children: searchResult.total }),
                    searchResult.total === 1 ? " result" : " results"
                  ] }) })
                ]
              }
            ) : /* @__PURE__ */ jsxs9("div", { className: "noResultsPane", children: [
              /* @__PURE__ */ jsx20("div", { className: "noResults", children: "No results" }),
              /* @__PURE__ */ jsx20("div", { className: "noResultsDetail", children: "Try different search terms" })
            ] }) }),
            hasQuery && !searchResult && searchError && /* @__PURE__ */ jsx20("div", { className: "noResultsPane", children: /* @__PURE__ */ jsx20("div", { className: "noResults", children: "Search error" }) })
          ] })
        }
      );
    } });
  }
};

// src/components/header.tsx
import { Fragment as Fragment5, jsx as jsx21, jsxs as jsxs10 } from "react/jsx-runtime";
function Header({
  block
}) {
  return /* @__PURE__ */ jsx21("header", { className: "notion-header", children: /* @__PURE__ */ jsxs10("div", { className: "notion-nav-header", children: [
    /* @__PURE__ */ jsx21(Breadcrumbs, { block }),
    /* @__PURE__ */ jsx21(Search, { block })
  ] }) });
}
function Breadcrumbs({
  block,
  rootOnly = false
}) {
  const { recordMap, mapPageUrl, components } = useNotionContext();
  const breadcrumbs = React13.useMemo(() => {
    const tempBreadcrumbs = getPageBreadcrumbs(recordMap, block.id);
    if (rootOnly) {
      return [tempBreadcrumbs == null ? void 0 : tempBreadcrumbs[0]].filter(Boolean);
    }
    return tempBreadcrumbs;
  }, [recordMap, block.id, rootOnly]);
  return /* @__PURE__ */ jsx21("div", { className: "breadcrumbs", children: breadcrumbs == null ? void 0 : breadcrumbs.map((breadcrumb, index) => {
    if (!breadcrumb) {
      return null;
    }
    const pageLinkProps = {};
    const componentMap = {
      pageLink: components.PageLink
    };
    if (breadcrumb.active) {
      componentMap.pageLink = (props) => /* @__PURE__ */ jsx21("div", { ...props });
    } else {
      pageLinkProps.href = mapPageUrl(breadcrumb.pageId);
    }
    return /* @__PURE__ */ jsxs10(React13.Fragment, { children: [
      /* @__PURE__ */ jsxs10(
        componentMap.pageLink,
        {
          className: cs("breadcrumb", breadcrumb.active && "active"),
          ...pageLinkProps,
          children: [
            breadcrumb.icon && /* @__PURE__ */ jsx21(PageIcon, { className: "icon", block: breadcrumb.block }),
            breadcrumb.title && /* @__PURE__ */ jsx21("span", { className: "title", children: breadcrumb.title })
          ]
        }
      ),
      index < breadcrumbs.length - 1 && /* @__PURE__ */ jsx21("span", { className: "spacer", children: "/" })
    ] }, breadcrumb.pageId);
  }) }, "breadcrumbs");
}
function Search({
  block,
  search,
  title = "Search"
}) {
  const { searchNotion, rootPageId, isShowingSearch, onHideSearch } = useNotionContext();
  const onSearchNotion = search || searchNotion;
  const [isSearchOpen, setIsSearchOpen] = React13.useState(isShowingSearch);
  React13.useEffect(() => {
    setIsSearchOpen(isShowingSearch);
  }, [isShowingSearch]);
  const onOpenSearch = React13.useCallback(() => {
    setIsSearchOpen(true);
  }, []);
  const onCloseSearch = React13.useCallback(() => {
    setIsSearchOpen(false);
    if (onHideSearch) {
      onHideSearch();
    }
  }, [onHideSearch]);
  useHotkeys("cmd+p", (event) => {
    onOpenSearch();
    event.preventDefault();
    event.stopPropagation();
  });
  useHotkeys("cmd+k", (event) => {
    onOpenSearch();
    event.preventDefault();
    event.stopPropagation();
  });
  const hasSearch = !!onSearchNotion;
  return /* @__PURE__ */ jsxs10(Fragment5, { children: [
    hasSearch && /* @__PURE__ */ jsxs10(
      "div",
      {
        role: "button",
        className: cs("breadcrumb", "button", "notion-search-button"),
        onClick: onOpenSearch,
        children: [
          /* @__PURE__ */ jsx21(SearchIcon, { className: "searchIcon" }),
          title && /* @__PURE__ */ jsx21("span", { className: "title", children: title })
        ]
      }
    ),
    isSearchOpen && hasSearch && /* @__PURE__ */ jsx21(
      SearchDialog,
      {
        isOpen: isSearchOpen,
        rootBlockId: rootPageId || (block == null ? void 0 : block.id),
        onClose: onCloseSearch,
        searchNotion: onSearchNotion
      }
    )
  ] });
}

// src/next.tsx
import React14 from "react";
import isEqual from "react-fast-compare";
import { jsx as jsx22 } from "react/jsx-runtime";
var wrapNextImage = (NextImage) => {
  return React14.memo(function ReactNotionXNextImage({
    src,
    alt,
    width,
    height,
    className,
    fill,
    ...rest
  }) {
    if (fill === "undefined") {
      fill = !(width && height);
    }
    return /* @__PURE__ */ jsx22(
      NextImage,
      {
        className,
        src,
        alt,
        width: !fill && width && height ? width : void 0,
        height: !fill && width && height ? height : void 0,
        fill,
        ...rest
      }
    );
  }, isEqual);
};
var wrapNextLegacyImage = (NextLegacyImage) => {
  return React14.memo(function ReactNotionXNextLegacyImage({
    src,
    alt,
    width,
    height,
    className,
    style,
    layout,
    ...rest
  }) {
    if (!layout) {
      layout = width && height ? "intrinsic" : "fill";
    }
    return /* @__PURE__ */ jsx22(
      NextLegacyImage,
      {
        className,
        src,
        alt,
        width: layout === "intrinsic" && width,
        height: layout === "intrinsic" && height,
        objectFit: style == null ? void 0 : style.objectFit,
        objectPosition: style == null ? void 0 : style.objectPosition,
        layout,
        ...rest
      }
    );
  }, isEqual);
};
function wrapNextLink(NextLink) {
  return ({
    href,
    as,
    passHref,
    prefetch,
    replace,
    scroll,
    shallow,
    locale,
    ...linkProps
  }) => {
    return /* @__PURE__ */ jsx22(
      NextLink,
      {
        href,
        as,
        passHref,
        prefetch,
        replace,
        scroll,
        shallow,
        locale,
        legacyBehavior: true,
        children: /* @__PURE__ */ jsx22("a", { ...linkProps })
      }
    );
  };
}

// src/context.tsx
import { jsx as jsx23 } from "react/jsx-runtime";
function DefaultLink(props) {
  return /* @__PURE__ */ jsx23("a", { target: "_blank", rel: "noopener noreferrer", ...props });
}
var DefaultLinkMemo = React15.memo(DefaultLink);
function DefaultPageLink(props) {
  return /* @__PURE__ */ jsx23("a", { ...props });
}
var DefaultPageLinkMemo = React15.memo(DefaultPageLink);
function DefaultEmbed(props) {
  return /* @__PURE__ */ jsx23(AssetWrapper, { ...props });
}
var DefaultHeader = Header;
function dummyLink({ href, rel, target, title, ...rest }) {
  return /* @__PURE__ */ jsx23("span", { ...rest });
}
var dummyComponent = (name) => () => {
  console.warn(
    `Warning: using empty component "${name}" (you should override this in NotionRenderer.components)`
  );
  return null;
};
var dummyOverrideFn = (_, defaultValueFn) => defaultValueFn();
var defaultComponents = {
  Image: null,
  // disable custom images by default
  Link: DefaultLinkMemo,
  PageLink: DefaultPageLinkMemo,
  Checkbox,
  Callout: void 0,
  // use the built-in callout rendering by default
  Button: void 0,
  // use the built-in button rendering by default
  Code: dummyComponent("Code"),
  Equation: dummyComponent("Equation"),
  Collection: dummyComponent("Collection"),
  Property: void 0,
  // use the built-in property rendering by default
  propertyTextValue: dummyOverrideFn,
  propertySelectValue: dummyOverrideFn,
  propertyRelationValue: dummyOverrideFn,
  propertyFormulaValue: dummyOverrideFn,
  propertyTitleValue: dummyOverrideFn,
  propertyPersonValue: dummyOverrideFn,
  propertyFileValue: dummyOverrideFn,
  propertyCheckboxValue: dummyOverrideFn,
  propertyUrlValue: dummyOverrideFn,
  propertyEmailValue: dummyOverrideFn,
  propertyPhoneNumberValue: dummyOverrideFn,
  propertyNumberValue: dummyOverrideFn,
  propertyLastEditedTimeValue: dummyOverrideFn,
  propertyCreatedTimeValue: dummyOverrideFn,
  propertyDateValue: dummyOverrideFn,
  propertyAutoIncrementIdValue: dummyOverrideFn,
  Pdf: dummyComponent("Pdf"),
  Tweet: dummyComponent("Tweet"),
  Modal: dummyComponent("Modal"),
  Header: DefaultHeader,
  Embed: DefaultEmbed
};
var defaultNotionContext = {
  recordMap: {
    block: {},
    collection: {},
    collection_view: {},
    collection_query: {},
    notion_user: {},
    signed_urls: {}
  },
  components: defaultComponents,
  mapPageUrl: defaultMapPageUrl(),
  mapImageUrl: defaultMapImageUrl,
  searchNotion: void 0,
  isShowingSearch: false,
  onHideSearch: void 0,
  fullPage: false,
  darkMode: false,
  previewImages: false,
  forceCustomImages: false,
  showCollectionViewDropdown: true,
  linkTableTitleProperties: true,
  isLinkCollectionToUrlProperty: false,
  showTableOfContents: false,
  minTableOfContentsItems: 3,
  defaultPageIcon: null,
  defaultPageCover: null,
  defaultPageCoverPosition: 0.5,
  zoom: null
};
var ctx = React15.createContext(defaultNotionContext);
function NotionContextProvider({
  components: themeComponents = {},
  children,
  mapPageUrl,
  mapImageUrl,
  rootPageId,
  ...rest
}) {
  for (const key of Object.keys(rest)) {
    if (rest[key] === void 0) {
      delete rest[key];
    }
  }
  const wrappedThemeComponents = React15.useMemo(
    () => ({
      ...themeComponents
    }),
    [themeComponents]
  );
  if (wrappedThemeComponents.nextImage && wrappedThemeComponents.nextLegacyImage) {
    console.warn(
      "You should not pass both nextImage and nextLegacyImage. Only nextImage component will be used."
    );
    wrappedThemeComponents.Image = wrapNextImage(themeComponents.nextImage);
  } else if (wrappedThemeComponents.nextImage) {
    wrappedThemeComponents.Image = wrapNextImage(themeComponents.nextImage);
  } else if (wrappedThemeComponents.nextLegacyImage) {
    wrappedThemeComponents.Image = wrapNextLegacyImage(
      themeComponents.nextLegacyImage
    );
  }
  if (wrappedThemeComponents.nextLink) {
    wrappedThemeComponents.nextLink = wrapNextLink(themeComponents.nextLink);
  }
  for (const key of Object.keys(wrappedThemeComponents)) {
    if (!wrappedThemeComponents[key]) {
      delete wrappedThemeComponents[key];
    }
  }
  const value = React15.useMemo(
    () => ({
      ...defaultNotionContext,
      ...rest,
      rootPageId,
      mapPageUrl: mapPageUrl != null ? mapPageUrl : defaultMapPageUrl(rootPageId),
      mapImageUrl: mapImageUrl != null ? mapImageUrl : defaultMapImageUrl,
      components: { ...defaultComponents, ...wrappedThemeComponents }
    }),
    [mapImageUrl, mapPageUrl, wrappedThemeComponents, rootPageId, rest]
  );
  return /* @__PURE__ */ jsx23(ctx.Provider, { value, children });
}
var NotionContextConsumer = ctx.Consumer;
var useNotionContext = () => {
  return React15.useContext(ctx);
};

// src/components/button.tsx
import { jsx as jsx24 } from "react/jsx-runtime";
function Button({
  block,
  blockId,
  className
}) {
  var _a, _b, _c, _d, _e, _f;
  const { recordMap, mapPageUrl } = useNotionContext();
  const [isSuccess, setIsSuccess] = React16.useState(false);
  const [isLoading, setIsLoading] = React16.useState(false);
  if (!block || block.type !== "button") {
    console.warn("Invalid button block:", { block });
    return null;
  }
  const automationId = (_a = block.format) == null ? void 0 : _a.automation_id;
  const blockColor = ((_b = block.format) == null ? void 0 : _b.block_color) || "default";
  const title = (_c = block.properties) == null ? void 0 : _c.title;
  if (!automationId) {
    return /* @__PURE__ */ jsx24("div", { className: cs("notion-button-block", blockId), children: /* @__PURE__ */ jsx24(
      "button",
      {
        type: "button",
        className: cs("notion-button", `notion-${blockColor}`, className),
        children: title ? /* @__PURE__ */ jsx24(Text, { value: title, block }) : "Button"
      }
    ) });
  }
  const automation = (_e = (_d = recordMap.automation) == null ? void 0 : _d[automationId]) == null ? void 0 : _e.value;
  if (!automation) {
    const buttonText2 = title ? getTextContent2(title) : "Button";
    return /* @__PURE__ */ jsx24("div", { className: cs("notion-button-block", blockId), children: /* @__PURE__ */ jsx24(
      "button",
      {
        type: "button",
        className: cs("notion-button", `notion-${blockColor}`, className),
        children: buttonText2
      }
    ) });
  }
  const buttonText = ((_f = automation.properties) == null ? void 0 : _f.name) || (title ? getTextContent2(title) : "Button");
  const handleClick = async (e) => {
    e.preventDefault();
    if (isLoading) return;
    setIsSuccess(false);
    setIsLoading(true);
    try {
      await executeAction();
    } finally {
      setIsLoading(false);
    }
  };
  const executeAction = async () => {
    var _a2, _b2, _c2, _d2, _e2, _f2, _g, _h, _i, _j, _k;
    const firstActionId = (_a2 = automation.action_ids) == null ? void 0 : _a2[0];
    if (!firstActionId) {
      console.warn("No actions defined for automation:", automationId);
      return;
    }
    const actionData = (_c2 = (_b2 = recordMap.automation_action) == null ? void 0 : _b2[firstActionId]) == null ? void 0 : _c2.value;
    if (!actionData) {
      console.warn("No action data found for ID:", firstActionId);
      return;
    }
    switch (actionData.type) {
      case "open_page": {
        const target = (_d2 = actionData.config) == null ? void 0 : _d2.target;
        if ((target == null ? void 0 : target.type) === "url" && target.url) {
          window.open(target.url, "_blank", "noopener,noreferrer");
        } else if ((target == null ? void 0 : target.type) === "page" && target.pageId) {
          const pageUrl = mapPageUrl(target.pageId);
          if (pageUrl) {
            window.location.href = pageUrl;
          }
        }
        break;
      }
      case "send_webhook":
      case "http_request": {
        const webhookUrl = (_f2 = (_e2 = actionData.config) == null ? void 0 : _e2.url) == null ? void 0 : _f2.trim();
        if (webhookUrl) {
          try {
            if (typeof window !== "undefined" && "/api/webhook-proxy") {
              const pageBlockId = Object.keys(recordMap.block || {}).find(
                (id) => {
                  var _a3;
                  const b = (_a3 = recordMap.block[id]) == null ? void 0 : _a3.value;
                  return b && "type" in b && b.type === "page";
                }
              );
              const pageBlock = pageBlockId ? (_g = recordMap.block[pageBlockId]) == null ? void 0 : _g.value : null;
              if (!pageBlock) {
                console.warn("No page block found for webhook payload");
                return;
              }
              const payload = {
                source: {
                  type: "automation",
                  automation_id: automationId,
                  action_id: actionData.id,
                  event_id: crypto.randomUUID(),
                  user_id: pageBlock.created_by_id || "unknown",
                  attempt: 1
                },
                data: {
                  object: "page",
                  id: pageBlock.id,
                  created_time: new Date(pageBlock.created_time).toISOString(),
                  last_edited_time: new Date(
                    pageBlock.last_edited_time
                  ).toISOString(),
                  created_by: {
                    object: "user",
                    id: pageBlock.created_by_id || "unknown"
                  },
                  last_edited_by: {
                    object: "user",
                    id: pageBlock.last_edited_by_id || "unknown"
                  },
                  cover: ((_h = pageBlock.format) == null ? void 0 : _h.page_cover) ? {
                    type: "external",
                    external: {
                      url: pageBlock.format.page_cover.startsWith("/") ? `https://www.notion.so${pageBlock.format.page_cover}` : pageBlock.format.page_cover
                    }
                  } : null,
                  icon: ((_i = pageBlock.format) == null ? void 0 : _i.page_icon) ? {
                    type: "external",
                    external: { url: pageBlock.format.page_icon }
                  } : null,
                  parent: {
                    type: "workspace",
                    workspace: true
                  },
                  archived: !pageBlock.alive,
                  in_trash: false,
                  is_locked: false,
                  properties: {
                    title: {
                      id: "title",
                      type: "title",
                      title: ((_j = pageBlock.properties) == null ? void 0 : _j.title) ? pageBlock.properties.title.map((t) => {
                        const text = typeof t === "string" ? t : t[0] || "";
                        return {
                          type: "text",
                          text: { content: text, link: null },
                          annotations: {
                            bold: false,
                            italic: false,
                            strikethrough: false,
                            underline: false,
                            code: false,
                            color: "default"
                          },
                          plain_text: text,
                          href: null
                        };
                      }) : []
                    }
                  },
                  url: `https://www.notion.so/${pageBlock.id}`,
                  public_url: window.location.href,
                  request_id: crypto.randomUUID()
                }
              };
              const headers = {
                "Content-Type": "application/json"
              };
              if ((_k = actionData.config) == null ? void 0 : _k.customHeaders) {
                for (const header of actionData.config.customHeaders) {
                  headers[header.key] = header.value;
                }
              }
              console.log("Sending webhook:", {
                url: webhookUrl,
                payload,
                headers
              });
              const response = await fetch("/api/webhook-proxy", {
                method: "POST",
                headers: {
                  "Content-Type": "application/json"
                },
                body: JSON.stringify({ url: webhookUrl, payload, headers })
              });
              const result = await response.json();
              if (!response.ok || !result.success) {
                console.error("Webhook request failed:", result);
              } else {
                console.log("Webhook sent successfully:", result);
                setIsSuccess(true);
                setTimeout(() => setIsSuccess(false), 2e3);
              }
            } else {
              console.warn(
                "Webhook functionality requires /api/webhook-proxy endpoint"
              );
            }
          } catch (err) {
            console.error("Error sending webhook:", err);
          }
        }
        break;
      }
      default:
        console.warn("Unsupported action type:", actionData.type);
    }
  };
  return /* @__PURE__ */ jsx24("div", { className: cs("notion-button-block", blockId), children: /* @__PURE__ */ jsx24(
    "button",
    {
      type: "button",
      className: cs(
        "notion-button",
        `notion-${blockColor}`,
        isSuccess ? "notion-success" : "",
        className
      ),
      onClick: handleClick,
      title: buttonText,
      disabled: isLoading,
      children: buttonText
    }
  ) });
}
function getTextContent2(text) {
  return text.map((t) => typeof t === "string" ? t : t[0] || "").join("");
}

// src/renderer.tsx
import mediumZoom from "@fisch0920/medium-zoom";
import "notion-types";
import { getBlockValue as getBlockValue3 } from "notion-utils";
import * as React21 from "react";

// src/block.tsx
import {
  getBlockCollectionId,
  getBlockIcon as getBlockIcon2,
  getBlockParentPage as getBlockParentPage2,
  getBlockValue as getBlockValue2,
  getListNestingLevel,
  getListNumber,
  getListStyle,
  getPageTableOfContents,
  getTextContent as getTextContent3,
  uuidToId as uuidToId2
} from "notion-utils";
import React20 from "react";

// src/components/audio.tsx
import "notion-types";
import { jsx as jsx25 } from "react/jsx-runtime";
function Audio({
  block,
  className
}) {
  var _a, _b, _c;
  const { recordMap } = useNotionContext();
  let source = recordMap.signed_urls[block.id] || ((_c = (_b = (_a = block.properties) == null ? void 0 : _a.source) == null ? void 0 : _b[0]) == null ? void 0 : _c[0]);
  if (!source) {
    return null;
  }
  if (block.space_id) {
    const url = new URL(source);
    url.searchParams.set("spaceId", block.space_id);
    source = url.toString();
  }
  return /* @__PURE__ */ jsx25("div", { className: cs("notion-audio", className), children: /* @__PURE__ */ jsx25("audio", { controls: true, preload: "none", src: source }) });
}

// src/components/file.tsx
import "notion-types";

// src/icons/file-icon.tsx
import "react";
import { jsx as jsx26 } from "react/jsx-runtime";
function FileIcon(props) {
  const { className, ...rest } = props;
  return /* @__PURE__ */ jsx26("svg", { className, ...rest, viewBox: "0 0 30 30", children: /* @__PURE__ */ jsx26("path", { d: "M22,8v12c0,3.866-3.134,7-7,7s-7-3.134-7-7V8c0-2.762,2.238-5,5-5s5,2.238,5,5v12c0,1.657-1.343,3-3,3s-3-1.343-3-3V8h-2v12c0,2.762,2.238,5,5,5s5-2.238,5-5V8c0-3.866-3.134-7-7-7S6,4.134,6,8v12c0,4.971,4.029,9,9,9s9-4.029,9-9V8H22z" }) });
}

// src/components/file.tsx
import { jsx as jsx27, jsxs as jsxs11 } from "react/jsx-runtime";
function File({
  block,
  className
}) {
  var _a, _b, _c, _d, _e;
  const { components, recordMap } = useNotionContext();
  let source = recordMap.signed_urls[block.id] || ((_c = (_b = (_a = block.properties) == null ? void 0 : _a.source) == null ? void 0 : _b[0]) == null ? void 0 : _c[0]);
  if (!source) {
    return null;
  }
  if (block.space_id) {
    const url = new URL(source);
    url.searchParams.set("spaceId", block.space_id);
    source = url.toString();
  }
  return /* @__PURE__ */ jsx27("div", { className: cs("notion-file", className), children: /* @__PURE__ */ jsxs11(
    components.Link,
    {
      className: "notion-file-link",
      href: source,
      target: "_blank",
      rel: "noopener noreferrer",
      children: [
        /* @__PURE__ */ jsx27(FileIcon, { className: "notion-file-icon" }),
        /* @__PURE__ */ jsxs11("div", { className: "notion-file-info", children: [
          /* @__PURE__ */ jsx27("div", { className: "notion-file-title", children: /* @__PURE__ */ jsx27(Text, { value: ((_d = block.properties) == null ? void 0 : _d.title) || [["File"]], block }) }),
          ((_e = block.properties) == null ? void 0 : _e.size) && /* @__PURE__ */ jsx27("div", { className: "notion-file-size", children: /* @__PURE__ */ jsx27(Text, { value: block.properties.size, block }) })
        ] })
      ]
    }
  ) });
}

// src/components/google-drive.tsx
import "notion-types";
import { jsx as jsx28, jsxs as jsxs12 } from "react/jsx-runtime";
function GoogleDrive({
  block,
  className
}) {
  var _a;
  const { components, mapImageUrl } = useNotionContext();
  const properties = (_a = block.format) == null ? void 0 : _a.drive_properties;
  if (!properties) return null;
  let domain;
  try {
    const url = new URL(properties.url);
    domain = url.hostname;
  } catch (e) {
  }
  return /* @__PURE__ */ jsx28("div", { className: cs("notion-google-drive", className), children: /* @__PURE__ */ jsxs12(
    components.Link,
    {
      className: "notion-google-drive-link",
      href: properties.url,
      target: "_blank",
      rel: "noopener noreferrer",
      children: [
        /* @__PURE__ */ jsx28("div", { className: "notion-google-drive-preview", children: /* @__PURE__ */ jsx28(
          GracefulImage,
          {
            src: mapImageUrl(properties.thumbnail, block),
            alt: properties.title || "Google Drive Document",
            loading: "lazy"
          }
        ) }),
        /* @__PURE__ */ jsxs12("div", { className: "notion-google-drive-body", children: [
          properties.title && /* @__PURE__ */ jsx28("div", { className: "notion-google-drive-body-title", children: properties.title }),
          properties.icon && domain && /* @__PURE__ */ jsxs12("div", { className: "notion-google-drive-body-source", children: [
            properties.icon && /* @__PURE__ */ jsx28(
              "div",
              {
                className: "notion-google-drive-body-source-icon",
                style: {
                  backgroundImage: `url(${properties.icon})`
                }
              }
            ),
            domain && /* @__PURE__ */ jsx28("div", { className: "notion-google-drive-body-source-domain", children: domain })
          ] })
        ] })
      ]
    }
  ) });
}

// src/components/page-aside.tsx
var import_lodash2 = __toESM(require_lodash(), 1);
import { uuidToId } from "notion-utils";
import React18 from "react";
import { jsx as jsx29, jsxs as jsxs13 } from "react/jsx-runtime";
function PageAside({
  toc,
  activeSection,
  setActiveSection,
  pageAside,
  hasToc,
  hasAside,
  className
}) {
  const throttleMs = 100;
  const actionSectionScrollSpy = React18.useMemo(
    () => (0, import_lodash2.default)(() => {
      const sections = document.getElementsByClassName("notion-h");
      let prevBBox = null;
      let currentSectionId = activeSection;
      for (const section of sections) {
        if (!section || !(section instanceof Element)) continue;
        if (!currentSectionId) {
          currentSectionId = section.dataset.id;
        }
        const bbox = section.getBoundingClientRect();
        const prevHeight = prevBBox ? bbox.top - prevBBox.bottom : 0;
        const offset = Math.max(150, prevHeight / 4);
        if (bbox.top - offset < 0) {
          currentSectionId = section.dataset.id;
          prevBBox = bbox;
          continue;
        }
        break;
      }
      setActiveSection(currentSectionId);
    }, throttleMs),
    [
      // explicitly not taking a dependency on activeSection
      setActiveSection
    ]
  );
  React18.useEffect(() => {
    if (!hasToc) {
      return;
    }
    window.addEventListener("scroll", actionSectionScrollSpy);
    actionSectionScrollSpy();
    return () => {
      window.removeEventListener("scroll", actionSectionScrollSpy);
    };
  }, [hasToc, actionSectionScrollSpy]);
  if (!hasAside) {
    return null;
  }
  return /* @__PURE__ */ jsxs13("aside", { className: cs("notion-aside", className), children: [
    hasToc && /* @__PURE__ */ jsxs13("div", { className: "notion-aside-table-of-contents", children: [
      /* @__PURE__ */ jsx29("div", { className: "notion-aside-table-of-contents-header", children: "Table of Contents" }),
      /* @__PURE__ */ jsx29("nav", { className: "notion-table-of-contents", children: toc.map((tocItem) => {
        const id = uuidToId(tocItem.id);
        return /* @__PURE__ */ jsx29(
          "a",
          {
            href: `#${id}`,
            className: cs(
              "notion-table-of-contents-item",
              `notion-table-of-contents-item-indent-level-${tocItem.indentLevel}`,
              activeSection === id && "notion-table-of-contents-active-item"
            ),
            children: /* @__PURE__ */ jsx29(
              "span",
              {
                className: "notion-table-of-contents-item-body",
                style: {
                  display: "inline-block",
                  marginLeft: tocItem.indentLevel * 16
                },
                children: tocItem.text
              }
            )
          },
          id
        );
      }) })
    ] }),
    pageAside
  ] });
}

// src/components/sync-pointer-block.tsx
import "notion-types";
import { jsx as jsx30 } from "react/jsx-runtime";
function SyncPointerBlock({
  block,
  level
}) {
  var _a, _b;
  if (!block) {
    if (true) {
      console.warn("missing sync pointer block");
    }
    return null;
  }
  const syncPointerBlock = block;
  const referencePointerId = (_b = (_a = syncPointerBlock == null ? void 0 : syncPointerBlock.format) == null ? void 0 : _a.transclusion_reference_pointer) == null ? void 0 : _b.id;
  if (!referencePointerId) {
    return null;
  }
  return /* @__PURE__ */ jsx30(
    NotionBlockRenderer,
    {
      level,
      blockId: referencePointerId
    },
    referencePointerId
  );
}

// src/icons/link-icon.tsx
import "react";
import { jsx as jsx31 } from "react/jsx-runtime";
function LinkIcon(props) {
  const { className, ...rest } = props;
  return /* @__PURE__ */ jsx31(
    "svg",
    {
      className,
      ...rest,
      viewBox: "0 0 16 16",
      width: "16",
      height: "16",
      children: /* @__PURE__ */ jsx31(
        "path",
        {
          fillRule: "evenodd",
          d: "M7.775 3.275a.75.75 0 001.06 1.06l1.25-1.25a2 2 0 112.83 2.83l-2.5 2.5a2 2 0 01-2.83 0 .75.75 0 00-1.06 1.06 3.5 3.5 0 004.95 0l2.5-2.5a3.5 3.5 0 00-4.95-4.95l-1.25 1.25zm-4.69 9.64a2 2 0 010-2.83l2.5-2.5a2 2 0 012.83 0 .75.75 0 001.06-1.06 3.5 3.5 0 00-4.95 0l-2.5 2.5a3.5 3.5 0 004.95 4.95l1.25-1.25a.75.75 0 00-1.06-1.06l-1.25 1.25a2 2 0 01-2.83 0z"
        }
      )
    }
  );
}

// src/block.tsx
import { Fragment as Fragment6, jsx as jsx32, jsxs as jsxs14 } from "react/jsx-runtime";
var tocIndentLevelCache = {};
var pageCoverStyleCache = {};
function Block(props) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _A, _B, _C, _D, _E, _F, _G, _H, _I, _J, _K, _L, _M, _N, _O;
  const ctx2 = useNotionContext();
  const {
    components,
    fullPage,
    darkMode,
    recordMap,
    mapPageUrl,
    mapImageUrl,
    showTableOfContents,
    minTableOfContentsItems,
    defaultPageIcon,
    defaultPageCover,
    defaultPageCoverPosition
  } = ctx2;
  const [activeSection, setActiveSection] = React20.useState(null);
  const {
    block,
    children,
    level,
    className,
    bodyClassName,
    header,
    footer,
    pageHeader,
    pageFooter,
    pageTitle,
    pageAside,
    pageCover,
    hideBlockId,
    disableHeader
  } = props;
  if (!block) {
    return null;
  }
  if (level === 0 && block.type === "collection_view") {
    ;
    block.type = "collection_view_page";
  }
  const blockId = hideBlockId ? "notion-block" : `notion-block-${uuidToId2(block.id)}`;
  switch (block.type) {
    case "collection_view_page":
    // fallthrough
    case "page":
      if (level === 0) {
        const {
          page_icon = defaultPageIcon,
          page_cover = defaultPageCover,
          page_cover_position = defaultPageCoverPosition,
          page_full_width,
          page_small_text
        } = block.format || {};
        if (fullPage) {
          const properties = block.type === "page" ? block.properties : {
            title: (_a = getBlockValue2(
              recordMap.collection[getBlockCollectionId(block, recordMap)]
            )) == null ? void 0 : _a.name
          };
          const coverPosition = (1 - (page_cover_position || 0.5)) * 100;
          const pageCoverObjectPosition = `center ${coverPosition}%`;
          let pageCoverStyle = pageCoverStyleCache[pageCoverObjectPosition];
          if (!pageCoverStyle) {
            pageCoverStyle = pageCoverStyleCache[pageCoverObjectPosition] = {
              objectPosition: pageCoverObjectPosition
            };
          }
          const pageIcon = (_b = getBlockIcon2(block, recordMap)) != null ? _b : defaultPageIcon;
          const isPageIconUrl = pageIcon && isUrl(pageIcon);
          const toc = getPageTableOfContents(
            block,
            recordMap
          );
          const hasToc = showTableOfContents && toc.length >= minTableOfContentsItems;
          const hasAside = !!((hasToc || pageAside) && !page_full_width);
          const hasPageCover = !!(pageCover || page_cover);
          return /* @__PURE__ */ jsxs14(
            "div",
            {
              className: cs(
                "notion",
                "notion-app",
                darkMode ? "dark-mode" : "light-mode",
                blockId,
                className
              ),
              children: [
                /* @__PURE__ */ jsx32("div", { className: "notion-viewport" }),
                /* @__PURE__ */ jsxs14("div", { className: "notion-frame", children: [
                  !disableHeader && /* @__PURE__ */ jsx32(components.Header, { block }),
                  header,
                  /* @__PURE__ */ jsxs14("div", { className: "notion-page-scroller", children: [
                    hasPageCover && (pageCover != null ? pageCover : /* @__PURE__ */ jsx32("div", { className: "notion-page-cover-wrapper", children: /* @__PURE__ */ jsx32(
                      LazyImage,
                      {
                        src: mapImageUrl(page_cover, block),
                        alt: getTextContent3(properties == null ? void 0 : properties.title),
                        priority: true,
                        className: "notion-page-cover",
                        style: pageCoverStyle
                      }
                    ) })),
                    /* @__PURE__ */ jsxs14(
                      "main",
                      {
                        className: cs(
                          "notion-page",
                          hasPageCover ? "notion-page-has-cover" : "notion-page-no-cover",
                          page_icon ? "notion-page-has-icon" : "notion-page-no-icon",
                          isPageIconUrl ? "notion-page-has-image-icon" : "notion-page-has-text-icon",
                          "notion-full-page",
                          page_full_width && "notion-full-width",
                          page_small_text && "notion-small-text",
                          bodyClassName
                        ),
                        children: [
                          page_icon && /* @__PURE__ */ jsx32(
                            PageIcon,
                            {
                              block,
                              defaultIcon: defaultPageIcon,
                              inline: false
                            }
                          ),
                          pageHeader,
                          /* @__PURE__ */ jsx32("h1", { className: "notion-title", children: pageTitle != null ? pageTitle : /* @__PURE__ */ jsx32(Text, { value: properties == null ? void 0 : properties.title, block }) }),
                          (block.type === "collection_view_page" || block.type === "page" && block.parent_table === "collection") && /* @__PURE__ */ jsx32(components.Collection, { block, ctx: ctx2 }),
                          block.type !== "collection_view_page" && /* @__PURE__ */ jsxs14(
                            "div",
                            {
                              className: cs(
                                "notion-page-content",
                                hasAside && "notion-page-content-has-aside",
                                hasToc && "notion-page-content-has-toc"
                              ),
                              children: [
                                /* @__PURE__ */ jsx32("article", { className: "notion-page-content-inner", children }),
                                hasAside && /* @__PURE__ */ jsx32(
                                  PageAside,
                                  {
                                    toc,
                                    activeSection,
                                    setActiveSection,
                                    hasToc,
                                    hasAside,
                                    pageAside
                                  }
                                )
                              ]
                            }
                          ),
                          pageFooter
                        ]
                      }
                    ),
                    footer
                  ] })
                ] })
              ]
            }
          );
        } else {
          return /* @__PURE__ */ jsxs14(
            "main",
            {
              className: cs(
                "notion",
                darkMode ? "dark-mode" : "light-mode",
                "notion-page",
                page_full_width && "notion-full-width",
                page_small_text && "notion-small-text",
                blockId,
                className,
                bodyClassName
              ),
              children: [
                /* @__PURE__ */ jsx32("div", { className: "notion-viewport" }),
                pageHeader,
                (block.type === "collection_view_page" || block.type === "page" && block.parent_table === "collection") && /* @__PURE__ */ jsx32(components.Collection, { block, ctx: ctx2 }),
                block.type !== "collection_view_page" && children,
                pageFooter
              ]
            }
          );
        }
      } else {
        const blockColor = (_c = block.format) == null ? void 0 : _c.block_color;
        return /* @__PURE__ */ jsx32(
          components.PageLink,
          {
            className: cs(
              "notion-page-link",
              blockColor && `notion-${blockColor}`,
              blockId
            ),
            href: mapPageUrl(block.id),
            children: /* @__PURE__ */ jsx32(PageTitle, { block })
          }
        );
      }
    case "header":
    // fallthrough
    case "sub_header":
    // fallthrough
    case "sub_sub_header":
    // fallthrough
    case "header_4": {
      if (!block.properties) return null;
      const blockColor = (_d = block.format) == null ? void 0 : _d.block_color;
      const id = uuidToId2(block.id);
      const title = getTextContent3(block.properties.title) || `Notion Header ${id}`;
      let indentLevel = tocIndentLevelCache[block.id];
      let indentLevelClass;
      if (indentLevel === void 0) {
        const page = getBlockParentPage2(block, recordMap);
        if (page) {
          const toc = getPageTableOfContents(page, recordMap);
          const tocItem = toc.find((tocItem2) => tocItem2.id === block.id);
          if (tocItem) {
            indentLevel = tocItem.indentLevel;
            tocIndentLevelCache[block.id] = indentLevel;
          }
        }
      }
      if (indentLevel !== void 0) {
        indentLevelClass = `notion-h-indent-${indentLevel}`;
      }
      const isH1 = block.type === "header";
      const isH2 = block.type === "sub_header";
      const isH3 = block.type === "sub_sub_header";
      const isH4 = block.type === "header_4";
      const classNameStr = cs(
        isH1 && "notion-h notion-h1",
        isH2 && "notion-h notion-h2",
        isH3 && "notion-h notion-h3",
        isH4 && "notion-h notion-h4",
        blockColor && `notion-${blockColor}`,
        indentLevelClass,
        blockId
      );
      const innerHeader = /* @__PURE__ */ jsxs14("span", { children: [
        /* @__PURE__ */ jsx32("div", { id, className: "notion-header-anchor" }),
        !((_e = block.format) == null ? void 0 : _e.toggleable) && /* @__PURE__ */ jsx32("a", { className: "notion-hash-link", href: `#${id}`, title, children: /* @__PURE__ */ jsx32(LinkIcon, {}) }),
        /* @__PURE__ */ jsx32("span", { className: "notion-h-title", children: /* @__PURE__ */ jsx32(Text, { value: block.properties.title, block }) })
      ] });
      let headerBlock = null;
      if (isH1) {
        headerBlock = /* @__PURE__ */ jsx32("h2", { className: classNameStr, "data-id": id, children: innerHeader });
      } else if (isH2) {
        headerBlock = /* @__PURE__ */ jsx32("h3", { className: classNameStr, "data-id": id, children: innerHeader });
      } else if (isH3) {
        headerBlock = /* @__PURE__ */ jsx32("h4", { className: classNameStr, "data-id": id, children: innerHeader });
      } else {
        headerBlock = /* @__PURE__ */ jsx32("h5", { className: classNameStr, "data-id": id, children: innerHeader });
      }
      if ((_f = block.format) == null ? void 0 : _f.toggleable) {
        return /* @__PURE__ */ jsxs14("details", { className: cs("notion-toggle", blockId), children: [
          /* @__PURE__ */ jsx32("summary", { children: headerBlock }),
          /* @__PURE__ */ jsx32("div", { children })
        ] });
      } else {
        return headerBlock;
      }
    }
    case "divider":
      return /* @__PURE__ */ jsx32("hr", { className: cs("notion-hr", blockId) });
    case "text": {
      if (!block.properties && !((_g = block.content) == null ? void 0 : _g.length)) {
        return /* @__PURE__ */ jsx32("div", { className: cs("notion-blank", blockId), children: "\xA0" });
      }
      const blockColor = (_h = block.format) == null ? void 0 : _h.block_color;
      return /* @__PURE__ */ jsxs14(
        "div",
        {
          className: cs(
            "notion-text",
            blockColor && `notion-${blockColor}`,
            blockId
          ),
          children: [
            ((_i = block.properties) == null ? void 0 : _i.title) && /* @__PURE__ */ jsx32(Text, { value: block.properties.title, block }),
            children && /* @__PURE__ */ jsx32("div", { className: "notion-text-children", children })
          ]
        }
      );
    }
    case "bulleted_list":
    // fallthrough
    case "numbered_list": {
      const wrapList = (content, start2) => block.type === "bulleted_list" ? /* @__PURE__ */ jsx32("ul", { className: cs("notion-list", "notion-list-disc", blockId), children: content }) : /* @__PURE__ */ jsx32(
        "ol",
        {
          start: start2,
          className: cs("notion-list", "notion-list-numbered", blockId),
          style: block.type === "numbered_list" ? {
            listStyleType: getListStyle(
              getListNestingLevel(block.id, recordMap.block)
            )
          } : void 0,
          children: content
        }
      );
      let output = null;
      const isTopLevel = block.type !== ((_j = getBlockValue2(recordMap.block[block.parent_id])) == null ? void 0 : _j.type);
      const start = getListNumber(block.id, recordMap.block);
      if (block.content) {
        const listItem = block.properties ? /* @__PURE__ */ jsx32("li", { children: /* @__PURE__ */ jsx32(Text, { value: block.properties.title, block }) }) : null;
        if (block.type === "bulleted_list") {
          output = /* @__PURE__ */ jsxs14(Fragment6, { children: [
            listItem,
            /* @__PURE__ */ jsx32("ul", { className: cs("notion-list", "notion-list-disc", blockId), children })
          ] });
        } else {
          const nestingLevel = getListNestingLevel(block.id, recordMap.block);
          output = /* @__PURE__ */ jsxs14(Fragment6, { children: [
            listItem,
            /* @__PURE__ */ jsx32(
              "ol",
              {
                className: cs("notion-list", "notion-list-numbered", blockId),
                style: {
                  listStyleType: getListStyle(nestingLevel + 1)
                },
                children
              }
            )
          ] });
        }
      } else {
        output = block.properties ? /* @__PURE__ */ jsx32("li", { children: /* @__PURE__ */ jsx32(Text, { value: block.properties.title, block }) }) : null;
      }
      return isTopLevel ? wrapList(output, start) : output;
    }
    case "embed":
      return /* @__PURE__ */ jsx32(components.Embed, { blockId, block });
    case "replit":
    // fallthrough
    case "tweet":
    // fallthrough
    case "maps":
    // fallthrough
    case "pdf":
    // fallthrough
    case "figma":
    // fallthrough
    case "typeform":
    // fallthrough
    case "codepen":
    // fallthrough
    case "excalidraw":
    // fallthrough
    case "image":
    // fallthrough
    case "gist":
    // fallthrough
    case "video":
      return /* @__PURE__ */ jsx32(AssetWrapper, { blockId, block });
    case "drive": {
      const properties = (_k = block.format) == null ? void 0 : _k.drive_properties;
      if (!properties) {
        if ((_l = block.format) == null ? void 0 : _l.display_source) {
          return /* @__PURE__ */ jsx32(AssetWrapper, { blockId, block });
        }
      }
      return /* @__PURE__ */ jsx32(
        GoogleDrive,
        {
          block,
          className: blockId
        }
      );
    }
    case "audio":
      return /* @__PURE__ */ jsx32(Audio, { block, className: blockId });
    case "file":
      return /* @__PURE__ */ jsx32(File, { block, className: blockId });
    case "equation":
      return /* @__PURE__ */ jsx32(
        components.Equation,
        {
          block,
          inline: false,
          className: blockId
        }
      );
    case "code":
      return /* @__PURE__ */ jsx32(components.Code, { block });
    case "column_list":
      return /* @__PURE__ */ jsx32("div", { className: cs("notion-row", blockId), children });
    case "column": {
      const spacerWidth = `min(32px, 4vw)`;
      const ratio = ((_m = block.format) == null ? void 0 : _m.column_ratio) || 0.5;
      const parent = getBlockValue2(recordMap.block[block.parent_id]);
      const columns = ((_n = parent == null ? void 0 : parent.content) == null ? void 0 : _n.length) || Math.max(2, Math.ceil(1 / ratio));
      const width = `calc((100% - (${columns - 1} * ${spacerWidth})) * ${ratio})`;
      const style = { width };
      return /* @__PURE__ */ jsxs14(Fragment6, { children: [
        /* @__PURE__ */ jsx32("div", { className: cs("notion-column", blockId), style, children }),
        /* @__PURE__ */ jsx32("div", { className: "notion-spacer" })
      ] });
    }
    case "quote": {
      if (!block.properties) return null;
      const blockColor = (_o = block.format) == null ? void 0 : _o.block_color;
      return /* @__PURE__ */ jsxs14(
        "blockquote",
        {
          className: cs(
            "notion-quote",
            blockColor && `notion-${blockColor}`,
            blockId
          ),
          children: [
            /* @__PURE__ */ jsx32("div", { children: /* @__PURE__ */ jsx32(Text, { value: block.properties.title, block }) }),
            children
          ]
        }
      );
    }
    case "collection_view":
      return /* @__PURE__ */ jsx32(components.Collection, { block, className: blockId, ctx: ctx2 });
    case "callout":
      if (components.Callout) {
        return /* @__PURE__ */ jsx32(components.Callout, { block, className: blockId });
      } else {
        return /* @__PURE__ */ jsxs14(
          "div",
          {
            className: cs(
              "notion-callout",
              ((_p = block.format) == null ? void 0 : _p.block_color) && `notion-${(_q = block.format) == null ? void 0 : _q.block_color}_co`,
              blockId
            ),
            children: [
              /* @__PURE__ */ jsx32(PageIcon, { block, hideDefaultIcon: true }),
              /* @__PURE__ */ jsxs14("div", { className: "notion-callout-text", children: [
                /* @__PURE__ */ jsx32(Text, { value: (_r = block.properties) == null ? void 0 : _r.title, block }),
                children
              ] })
            ]
          }
        );
      }
    case "bookmark": {
      if (!block.properties) return null;
      const link = block.properties.link;
      if (!link || !((_s = link[0]) == null ? void 0 : _s[0])) return null;
      let title = getTextContent3(block.properties.title);
      if (!title) {
        title = getTextContent3(link);
      }
      if (title) {
        if (title.startsWith("http")) {
          try {
            const url = new URL(title);
            title = url.hostname;
          } catch (e) {
          }
        }
      }
      return /* @__PURE__ */ jsx32("div", { className: "notion-row", children: /* @__PURE__ */ jsxs14(
        components.Link,
        {
          target: "_blank",
          rel: "noopener noreferrer",
          className: cs(
            "notion-bookmark",
            ((_t = block.format) == null ? void 0 : _t.block_color) && `notion-${block.format.block_color}`,
            blockId
          ),
          href: link[0][0],
          children: [
            /* @__PURE__ */ jsxs14("div", { children: [
              title && /* @__PURE__ */ jsx32("div", { className: "notion-bookmark-title", children: /* @__PURE__ */ jsx32(Text, { value: [[title]], block }) }),
              ((_u = block.properties) == null ? void 0 : _u.description) && /* @__PURE__ */ jsx32("div", { className: "notion-bookmark-description", children: /* @__PURE__ */ jsx32(Text, { value: (_v = block.properties) == null ? void 0 : _v.description, block }) }),
              /* @__PURE__ */ jsxs14("div", { className: "notion-bookmark-link", children: [
                ((_w = block.format) == null ? void 0 : _w.bookmark_icon) && /* @__PURE__ */ jsx32("div", { className: "notion-bookmark-link-icon", children: /* @__PURE__ */ jsx32(
                  LazyImage,
                  {
                    src: mapImageUrl((_x = block.format) == null ? void 0 : _x.bookmark_icon, block),
                    alt: title
                  }
                ) }),
                /* @__PURE__ */ jsx32("div", { className: "notion-bookmark-link-text", children: /* @__PURE__ */ jsx32(Text, { value: link, block }) })
              ] })
            ] }),
            ((_y = block.format) == null ? void 0 : _y.bookmark_cover) && /* @__PURE__ */ jsx32("div", { className: "notion-bookmark-image", children: /* @__PURE__ */ jsx32(
              LazyImage,
              {
                src: mapImageUrl((_z = block.format) == null ? void 0 : _z.bookmark_cover, block),
                alt: getTextContent3((_A = block.properties) == null ? void 0 : _A.title),
                style: {
                  objectFit: "cover"
                }
              }
            ) })
          ]
        }
      ) });
    }
    case "toggle":
      return /* @__PURE__ */ jsxs14("details", { className: cs("notion-toggle", blockId), children: [
        /* @__PURE__ */ jsx32("summary", { children: /* @__PURE__ */ jsx32(Text, { value: (_B = block.properties) == null ? void 0 : _B.title, block }) }),
        /* @__PURE__ */ jsx32("div", { children })
      ] });
    case "button": {
      const ButtonComponent = components.Button || Button;
      return /* @__PURE__ */ jsx32(
        ButtonComponent,
        {
          blockId,
          block,
          className: blockId
        }
      );
    }
    case "table_of_contents": {
      const page = getBlockParentPage2(block, recordMap);
      if (!page) return null;
      const toc = getPageTableOfContents(page, recordMap);
      const blockColor = (_C = block.format) == null ? void 0 : _C.block_color;
      return /* @__PURE__ */ jsx32(
        "div",
        {
          className: cs(
            "notion-table-of-contents",
            blockColor && `notion-${blockColor}`,
            blockId
          ),
          children: toc.map((tocItem) => /* @__PURE__ */ jsx32(
            "a",
            {
              href: `#${uuidToId2(tocItem.id)}`,
              className: "notion-table-of-contents-item",
              children: /* @__PURE__ */ jsx32(
                "span",
                {
                  className: "notion-table-of-contents-item-body",
                  style: {
                    display: "inline-block",
                    marginLeft: tocItem.indentLevel * 24
                  },
                  children: tocItem.text
                }
              )
            },
            tocItem.id
          ))
        }
      );
    }
    case "to_do": {
      const isChecked = ((_F = (_E = (_D = block.properties) == null ? void 0 : _D.checked) == null ? void 0 : _E[0]) == null ? void 0 : _F[0]) === "Yes";
      return /* @__PURE__ */ jsxs14("div", { className: cs("notion-to-do", blockId), children: [
        /* @__PURE__ */ jsxs14("div", { className: "notion-to-do-item", children: [
          /* @__PURE__ */ jsx32(components.Checkbox, { blockId, isChecked }),
          /* @__PURE__ */ jsx32(
            "div",
            {
              className: cs(
                "notion-to-do-body",
                isChecked && `notion-to-do-checked`
              ),
              children: /* @__PURE__ */ jsx32(Text, { value: (_G = block.properties) == null ? void 0 : _G.title, block })
            }
          )
        ] }),
        /* @__PURE__ */ jsx32("div", { className: "notion-to-do-children", children })
      ] });
    }
    case "transclusion_container":
      return /* @__PURE__ */ jsx32("div", { className: cs("notion-sync-block", blockId), children });
    case "transclusion_reference":
      return /* @__PURE__ */ jsx32(SyncPointerBlock, { ...props, level: level + 1 });
    case "alias": {
      const blockPointerId = (_I = (_H = block == null ? void 0 : block.format) == null ? void 0 : _H.alias_pointer) == null ? void 0 : _I.id;
      const linkedBlock = getBlockValue2(recordMap.block[blockPointerId]);
      if (!linkedBlock) {
        console.log('"alias" missing block', blockPointerId);
        return null;
      }
      return /* @__PURE__ */ jsx32(
        components.PageLink,
        {
          className: cs("notion-page-link", blockPointerId),
          href: mapPageUrl(blockPointerId),
          children: /* @__PURE__ */ jsx32(PageTitle, { block: linkedBlock })
        }
      );
    }
    case "table":
      return /* @__PURE__ */ jsx32("table", { className: cs("notion-simple-table", blockId), children: /* @__PURE__ */ jsx32("tbody", { children }) });
    case "table_row": {
      const tableBlock = getBlockValue2(
        recordMap.block[block.parent_id]
      );
      if (!tableBlock) {
        return null;
      }
      const order = (_J = tableBlock.format) == null ? void 0 : _J.table_block_column_order;
      const formatMap = (_K = tableBlock.format) == null ? void 0 : _K.table_block_column_format;
      const backgroundColor = (_L = block.format) == null ? void 0 : _L.block_color;
      const hasRowHeader = ((_M = tableBlock.format) == null ? void 0 : _M.table_block_column_header) === true;
      const hasColumnHeader = ((_N = tableBlock.format) == null ? void 0 : _N.table_block_row_header) === true;
      const isHeaderRow = hasRowHeader && ((_O = tableBlock.content) == null ? void 0 : _O[0]) === block.id;
      if (!tableBlock || !order) {
        return null;
      }
      return /* @__PURE__ */ jsx32(
        "tr",
        {
          className: cs(
            "notion-simple-table-row",
            backgroundColor && `notion-${backgroundColor}`,
            isHeaderRow && "notion-simple-table-header-row",
            blockId
          ),
          children: order.map((column, columnIndex) => {
            var _a2, _b2, _c2;
            const color = (_a2 = formatMap == null ? void 0 : formatMap[column]) == null ? void 0 : _a2.color;
            const isHeaderColumn = hasColumnHeader && columnIndex === 0;
            return /* @__PURE__ */ jsx32(
              "td",
              {
                className: cs(
                  color ? `notion-${color}` : "",
                  isHeaderColumn && "notion-simple-table-header-cell"
                ),
                style: {
                  width: ((_b2 = formatMap == null ? void 0 : formatMap[column]) == null ? void 0 : _b2.width) || 120
                },
                children: /* @__PURE__ */ jsx32("div", { className: "notion-simple-table-cell", children: /* @__PURE__ */ jsx32(
                  Text,
                  {
                    value: ((_c2 = block.properties) == null ? void 0 : _c2[column]) || [["\u3164"]],
                    block
                  }
                ) })
              },
              column
            );
          })
        }
      );
    }
    case "external_object_instance":
      return /* @__PURE__ */ jsx32(EOI, { block, className: blockId });
    default:
      if (true) {
        console.log(
          "Unsupported block type " + block.type,
          JSON.stringify(block, null, 2)
        );
      }
      return /* @__PURE__ */ jsx32("div", {});
  }
}

// src/renderer.tsx
import { jsx as jsx33 } from "react/jsx-runtime";
function NotionRenderer({
  components,
  recordMap,
  mapPageUrl,
  mapImageUrl,
  searchNotion,
  isShowingSearch,
  onHideSearch,
  fullPage,
  rootPageId,
  rootDomain,
  darkMode,
  previewImages,
  forceCustomImages,
  showCollectionViewDropdown,
  linkTableTitleProperties,
  isLinkCollectionToUrlProperty,
  isImageZoomable = true,
  showTableOfContents,
  minTableOfContentsItems,
  defaultPageIcon,
  defaultPageCover,
  defaultPageCoverPosition,
  ...rest
}) {
  const zoom = React21.useMemo(
    () => !!globalThis.window && mediumZoom({
      background: "rgba(0, 0, 0, 0.8)",
      minZoomScale: 2,
      margin: getMediumZoomMargin()
    }),
    []
  );
  return /* @__PURE__ */ jsx33(
    NotionContextProvider,
    {
      components,
      recordMap,
      mapPageUrl,
      mapImageUrl,
      searchNotion,
      isShowingSearch,
      onHideSearch,
      fullPage,
      rootPageId,
      rootDomain,
      darkMode,
      previewImages,
      forceCustomImages,
      showCollectionViewDropdown,
      linkTableTitleProperties,
      isLinkCollectionToUrlProperty,
      showTableOfContents,
      minTableOfContentsItems,
      defaultPageIcon,
      defaultPageCover,
      defaultPageCoverPosition,
      zoom: isImageZoomable ? zoom : null,
      children: /* @__PURE__ */ jsx33(NotionBlockRenderer, { ...rest })
    }
  );
}
function NotionBlockRenderer({
  level = 0,
  blockId,
  ...props
}) {
  var _a;
  const { recordMap } = useNotionContext();
  const id = blockId || Object.keys(recordMap.block)[0];
  const block = getBlockValue3(recordMap.block[id]);
  if (!block) {
    if (true) {
      console.warn("missing block", blockId);
    }
    return null;
  }
  return /* @__PURE__ */ jsx33(Block, { level, block, ...props, children: (_a = block == null ? void 0 : block.content) == null ? void 0 : _a.map((contentBlockId) => /* @__PURE__ */ jsx33(
    NotionBlockRenderer,
    {
      blockId: contentBlockId,
      level: level + 1,
      ...props
    },
    contentBlockId
  )) }, id);
}
function getMediumZoomMargin() {
  const width = window.innerWidth;
  if (width < 500) {
    return 8;
  } else if (width < 800) {
    return 20;
  } else if (width < 1280) {
    return 30;
  } else if (width < 1600) {
    return 40;
  } else if (width < 1920) {
    return 48;
  } else {
    return 72;
  }
}
export {
  Breadcrumbs,
  Button,
  Header,
  NotionContextConsumer,
  NotionContextProvider,
  NotionRenderer,
  PageIcon,
  PageIconImpl,
  Search,
  Text,
  cs,
  dummyLink,
  formatDate,
  formatNotionDateTime,
  getHashFragmentValue,
  getUrlParams,
  getYoutubeId,
  isBrowser,
  isUrl,
  useNotionContext
};
